<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daianna Gonzalez-Padilla">
<meta name="dcterms.date" content="2024-08-14">
<meta name="description" content="A complete and interpretable explanation of the Fisher’s exact test in the context of gene set-based enrichment analysis. The genomic applicability of this test is exemplified with a functional enrichment analysis for DEGs.">

<title>Daianna Gonzalez-Padilla - Fisher’s exact test for enrichment analysis of gene sets</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/logo-removebg.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Daianna Gonzalez-Padilla</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">About me</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../materials.qmd"> 
<span class="menu-text">Other materials</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://daianna21.github.io"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="http://www.linkedin.com/in/daianna-glez"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://x.com/daianna_glez/"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:glezdaianna@gmail.com"> <i class="bi bi-envelope-at" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../blog.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#what-youll-learn-here" id="toc-what-youll-learn-here" class="nav-link" data-scroll-target="#what-youll-learn-here">What you’ll learn here</a></li>
  <li><a href="#the-fishers-exact-test" id="toc-the-fishers-exact-test" class="nav-link" data-scroll-target="#the-fishers-exact-test">The Fisher’s exact test</a>
  <ul class="collapse">
  <li><a href="#creating-and-understanding-contingency-tables" id="toc-creating-and-understanding-contingency-tables" class="nav-link" data-scroll-target="#creating-and-understanding-contingency-tables">1. Creating and understanding contingency tables</a></li>
  <li><a href="#testing-for-gene-set-enrichment" id="toc-testing-for-gene-set-enrichment" class="nav-link" data-scroll-target="#testing-for-gene-set-enrichment">2. Testing for gene-set enrichment</a>
  <ul class="collapse">
  <li><a href="#one-sided-test" id="toc-one-sided-test" class="nav-link" data-scroll-target="#one-sided-test"><strong>One-sided test</strong></a></li>
  </ul></li>
  <li><a href="#assessing-gene-set-depletion" id="toc-assessing-gene-set-depletion" class="nav-link" data-scroll-target="#assessing-gene-set-depletion">Assessing gene-set depletion?</a></li>
  <li><a href="#two-sided-test" id="toc-two-sided-test" class="nav-link" data-scroll-target="#two-sided-test"><strong>Two-sided test</strong></a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#relevant-links" id="toc-relevant-links" class="nav-link" data-scroll-target="#relevant-links">Relevant links</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Fisher’s exact test for enrichment analysis of gene sets</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Fisher’s exact test</div>
    <div class="quarto-category">overrepresentation</div>
    <div class="quarto-category">ORA</div>
    <div class="quarto-category">enrichment</div>
    <div class="quarto-category">Gene Set Enrichment Analysis</div>
    <div class="quarto-category">GO</div>
    <div class="quarto-category">KEGG</div>
    <div class="quarto-category">Functional enrichment</div>
  </div>
  </div>

<div>
  <div class="description">
    <p>A complete and interpretable explanation of the Fisher’s exact test in the context of gene set-based enrichment analysis. The genomic applicability of this test is exemplified with a functional enrichment analysis for DEGs.</p>
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Daianna Gonzalez-Padilla </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 14, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>Imagine you have analyzed high-throughput omics data and have ended up with a list of candidate/interest genes. These genes could be, for instance, differently expressed genes (DEGs), genes associated with a certain disease, genes affected by risk SNPs, markers for cell types, etc. Any list of <strong>experimentally-derived genes</strong> with biological or clinical relevance applies.</p>
<p>Now, what comes after that? If you found a few genes, you can do a little of research about them and make sense of your results, but what if you got thousands of genes? That’s not an easy task anymore and becomes unscalable. Moreover, despite each individual found gene providing valuable information, we can reveal more definite insights by studying all genes at the same time.</p>
<p>So what’s to do? We need to start thinking of these <strong>genes as a unique <u>set</u></strong> with its own properties, and ask questions such as <em>what do these genes have in common?</em> <em>what makes them being involved or affected by the condition under study</em>, and <em>what does that imply for cell functionality?</em></p>
<p><strong>Functional enrichment analysis</strong> based on the <span style="background-color: #FFF68F"><strong>one-sided Fisher’s exact test</strong></span> aims to test whether <span style="color: #EE3A8C"><strong>our genes <u>as a set</u></strong></span> are associated with certain cellular properties or functions by assessing if <em>sets of known genes that share biological activities or attributes</em> (<span style="color: #00C5CD"><strong>functional gene sets</strong></span>), have a statistically significant <strong>large overlap</strong> with our genes, i.e., if such functional gene sets are, what we call <u><strong>over-represented</strong></u> (or <u><strong>enriched</strong></u>, which is the same), among our genes. Stay on this post to learn more about how this test operates and how to understand and apply it to analyze sets of genes.</p>
</section>
<section id="what-youll-learn-here" class="level1">
<h1>What you’ll learn here</h1>
<ol type="1">
<li>Comprehend the purpose and the basics of a functional enrichment analysis based on gene sets.</li>
<li>Understand the probabilistic and statistic foundations of the Fisher’s exact test and interpret it with gene sets.</li>
<li>Clarify the required input gene sets to correctly perform this analysis.</li>
<li>Learn how to run this test using R.</li>
<li>Differentiate <strong>enrichment</strong> vs.&nbsp;<strong>depletion</strong> when performing this test.</li>
</ol>
</section>
<section id="the-fishers-exact-test" class="level1">
<h1>The Fisher’s exact test</h1>
<p>This is a classical statistical test based on the hypergeometric distribution (thus also called <strong>hypergeometric test</strong>) used to assess if there is a significant association between two categorical binary variables, such as differential expression of the genes and their involvement in a specific biological process. The association of the two variables is calculated in terms of the <strong>intersection size</strong> between their elements.</p>
<p>For illustrative purposes to show the application of this test for gene set-based functional enrichment analysis, imagine we have run a differential gene expression analysis (DGE) for a condition and have obtained a list of DEGs. As part of downstream analyses, we want to know how many of such DEGs participate in a specific biological process of interest and if that number significantly deviates from what is expected by chance.</p>
<p>In the next sections we’ll go into more detail on how to demarcate the required sets of genes, represent them in contingency tables, assess enrichment (or depletion) in R, and how to interpret this test and its results.</p>
<section id="creating-and-understanding-contingency-tables" class="level2">
<h2 class="anchored" data-anchor-id="creating-and-understanding-contingency-tables">1. Creating and understanding contingency tables</h2>
<p>The following are the required gene sets and their specifications. Please note that when referring to a set of genes, we are specifically referring to the set of gene IDs; that all we need to run this test!</p>
<ul>
<li><p><span style="color: #595959"><strong>Gene universe/background</strong></span>: these are all the genes that were experimentally measured and interrogated to generate the list of interest genes (our gene set). In our hypothetical scenario this is given by all genes that were assessed for DGE.</p>
<p>→ Let’s denote as “DE” the variable defining if a gene is differentially expressed (<span style="color: #EE3A8C"><strong>DEG</strong></span>) or not (<span style="color: #D8BFD8"><strong>non-DEG</strong></span>).</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How to find my gene universe?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since we are testing enrichment of functional gene sets among our DEGs, and not the other way around, our “gene universe” must correspond to the total number of genes for which we assessed DGE. Within those genes, categorized by DE, we search for genes contained in a functional set.</p>
</div>
</div></li>
<li><p><span style="color: #EE3A8C"><strong>Experimentally-derived gene set</strong></span> (our gene set): a list of interest genes derived from our experimental measurements and analyses, and to which we want to assign biological meanings. In our example case this is the list of DEGs from the <span style="color: #595959"><strong>gene universe</strong></span>.</p></li>
<li><p><span style="color: #00C5CD"><strong>Functional gene set</strong></span>: a set of genes with shared biological attributes (e.g.&nbsp;participation in a pathway, implication in the same biological process, sharing a molecular activity, or if their gene products are located in a certain cell compartment). Conveniently, functional gene sets are accessible via databases such as <a href="https://geneontology.org">Gene Ontology (GO)</a> <span class="citation" data-cites="ashburner2000">(<a href="#ref-ashburner2000" role="doc-biblioref">Ashburner et al. 2000</a>)</span> and the <a href="https://www.kegg.jp">Kyoto encyclopedia of genes and genomes (KEGG)</a> <span class="citation" data-cites="kanehisa2000">(<a href="#ref-kanehisa2000" role="doc-biblioref">Kanehisa 2000</a>)</span>.</p>
<p>→ Let’s denote as “BP” the variable determining if a gene is involved in the biological process (<span style="color: #00C5CD"><strong>BP+</strong></span>) or not (<span style="color: #C1CDCD"><strong>BP-</strong></span>).</p></li>
</ul>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>There may be cases where you have some genes in your functional set that are not present in the gene universe. Since they were not tested for DGE we can’t categorize them as DEGs or non-DEGs. Despite this, given that most genes are non-DEGs and we often assume that, you can either consider them non-DEGs or exclude them from the analysis to avoid introducing ambiguous information.</p>
</div>
</div>
<p>Fisher’s exact test aims to detect functional gene sets whose genes are unusually represented (<strong>enriched</strong>/<strong>over-represented</strong> or <strong>under-represented</strong>) in our set of interest genes. That is, if there’s a large (or small) overlap between the genes of the functional set (BP+ genes in this case) and the genes we classified as relevant from our data (DEGs in this case).</p>
<p>Following our hypothetical example, let’s suppose we have analyzed DGE for a total of 32 genes and of these 15 were <span style="color: #EE3A8C"><strong>DEGs</strong></span>. Suppose we have a functional gene set containing 12 genes involved in a biological process (<span style="color: #00C5CD"><strong>BP+</strong></span>) and of these 9 were DEGs. Having identified such gene sets, we proceed to create a <strong>2x2 contingency table</strong> showing the number of genes that belong to each category for each variable.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/contingency_table_3.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="421"></p>
</figure>
</div>
<p>The numbers of <span style="color: #EE3A8C"><strong>DEGs</strong></span>, <span style="color: #D8BFD8"><strong>non-DEG</strong></span>, <span style="color: #00C5CD"><strong>BP+</strong></span> genes, and <span style="color: #C1CDCD"><strong>BP-</strong></span> genes are called the <em>marginal values</em>, as they lie on the periphery of this 2x2 table. The middle cells (<span class="math inline">\(a\)</span>’s) are the <em>joint values</em>, as they contain the overlap between the categories of the two variables.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that by changing only one joint value, say <span class="math inline">\(a_{11}\)</span>, all the remaining joint values will be already determined since the marginals are fixed. In that way we can describe whole contingency tables with only their <span class="math inline">\(a_{11}\)</span> corner values.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define example gene universe: 32 total genes analyzed </span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>gene_universe <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Srsf5"</span>, <span class="st">"Gm15387"</span>, <span class="st">"Mprip"</span>, <span class="st">"Pim1"</span>, <span class="st">"Bnip3l"</span>, <span class="st">"Efr3a"</span>, <span class="st">"Marco"</span>, <span class="st">"Tuba1a"</span>, <span class="st">"Gsk3a"</span>,</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"Dap3"</span>, <span class="st">"Tmod3"</span>, <span class="st">"Dnajb1"</span>, <span class="st">"Tulp4"</span>, <span class="st">"Lsm14b"</span>, <span class="st">"Khdrbs2"</span>, <span class="st">"Mfsd9"</span>, <span class="st">"Ufd1"</span>, <span class="st">"Ypel5"</span>,</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"Rbm3"</span>, <span class="st">"Dnajc28"</span>, <span class="st">"Msrb2"</span>, <span class="st">"Memo1"</span>, <span class="st">"Cebpg"</span>, <span class="st">"Flywch1"</span>, <span class="st">"Ip6k1"</span>, <span class="st">"Nudt4"</span>, <span class="st">"Csnk1e"</span>,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"Qsox2"</span>, <span class="st">"Gm43517"</span>, <span class="st">"Rhbdf2"</span>, <span class="st">"Fbxw11"</span>, <span class="st">"Hmgxb4"</span>) </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(gene_universe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define 15 DEGs from universe</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">08122024</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>DEGs <span class="ot">&lt;-</span> <span class="fu">sample</span>(gene_universe, <span class="dv">15</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>DEGs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Srsf5"   "Flywch1" "Ip6k1"   "Cebpg"   "Nudt4"   "Ypel5"   "Marco"  
 [8] "Csnk1e"  "Qsox2"   "Gsk3a"   "Dap3"    "Tuba1a"  "Efr3a"   "Hmgxb4" 
[15] "Gm15387"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(DEGs))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Remaining 17 non-DEGs</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>non_DEGs <span class="ot">&lt;-</span> gene_universe[<span class="sc">!</span>gene_universe <span class="sc">%in%</span> DEGs]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Functional set of 12 BP+ genes (with 9 being DEGs)</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>funct_gene_set <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">sample</span>(DEGs, <span class="dv">9</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>), <span class="fu">sample</span>(non_DEGs, <span class="dv">3</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>))</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(funct_gene_set))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="do">## BP- genes</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>genes_not_func_set <span class="ot">&lt;-</span> gene_universe[<span class="sc">!</span>gene_universe <span class="sc">%in%</span> funct_gene_set]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Intersections:</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. DEGs and BP+ genes</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>DEGs_in_set <span class="ot">&lt;-</span> <span class="fu">intersect</span>(DEGs, funct_gene_set)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. non-DEGs and BP+ genes</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>non_DEGs_in_set <span class="ot">&lt;-</span> <span class="fu">intersect</span>(non_DEGs, funct_gene_set)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. DEGs and BP- genes</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>DEGs_not_in_set <span class="ot">&lt;-</span> <span class="fu">intersect</span>(DEGs, genes_not_func_set)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. non-DEGs and BP- genes</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>non_DEGs_not_in_set <span class="ot">&lt;-</span> <span class="fu">intersect</span>(non_DEGs, genes_not_func_set)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="do">## Create contingency table</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fu">length</span>(DEGs_in_set), <span class="fu">length</span>(DEGs_not_in_set), </span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>            <span class="fu">length</span>(non_DEGs_in_set), <span class="fu">length</span>(non_DEGs_not_in_set)), </span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>            <span class="at">byrow =</span> <span class="cn">TRUE</span>, <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>m</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    9    6
[2,]    3   14</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>When we test for “association” between DE and BP note that it could be either <strong>positive</strong> or <strong>negative</strong>, with the former implying a larger overlap between DEGs and BP+ genes, and thus corresponding to <strong>enrichment</strong>, and the latter implying a smaller overlap and thus meaning <strong>depletion</strong>. We’ll also delve into the statistical implications of each type of association in the next section.</p>
</div>
</div>
</section>
<section id="testing-for-gene-set-enrichment" class="level2">
<h2 class="anchored" data-anchor-id="testing-for-gene-set-enrichment">2. Testing for gene-set enrichment</h2>
<p>The Fisher’s exact test is used to determine the <span style="background-color: #FFFFE0">probability of observing the joint value <span class="math inline">\(a_{11}\)</span> in our table or a more extreme value</span> (for contingency tables with the same fixed marginal totals) under the <strong>null hypothesis (</strong><span class="math inline">\(H_0\)</span><strong>)</strong> that there’s <u>no association between the categorical variables</u>, i.e., that they are independent. These probabilities are referred to as <em>p</em>-values. Smaller probabilities offer strong evidence against the null hypothesis, which is translated into statistical evidence to accept the <strong>alternative hypothesis (</strong><span class="math inline">\(H_A\)</span><strong>)</strong> of <u>association between the two variables</u>.</p>
<p>This is how it operates:</p>
<ol type="1">
<li>Given the fixed marginal totals, we first we compute all possible contingency tables by enumerating all possible joint values <span class="math inline">\(a_{11}\)</span>: in this case from 0 to 12, which are the minimum and maximum number of genes that can be both <span style="color: #EE3A8C"><strong>DEGs</strong></span> and <span style="color: #00C5CD"><strong>BP+</strong></span>, respectively.</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/all_tables.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin: 0px" width="449"></p>
</figure>
</div>
<div style="background-color:#F2F2F2; padding: 15px; margin-bottom:0px">
<div class="line-block">❗️ The top left cell (<span class="math inline">\(a_{11}\)</span>) of each table always corresponds to the joint value positive for both variables under examination (<span style="color: #00C5CD"><strong>BP+</strong></span> <span style="color: #EE3A8C"><strong>DEGs</strong></span>).</div>
</div>
<p><img src="images/white_space.png" style="margin: 0px" width="30" height="4"></p>
<ol start="2" type="1">
<li><p>Then we estimate the probability of having such joint values in each contingency table under the null hypothesis. To do that we need to calculate the number of ways of obtaining the joint values by randomness in the gene selection process. Only 3 values need to be estimated for that:</p>
<ol type="1">
<li><p>The ways of selecting <span class="math inline">\(a_{11} + a_{12}\)</span> = 15 genes (corresponding to the number of <span style="color: #EE3A8C"><strong>DEGs</strong></span>) without replacement from the total genes in the <span style="color: #595959"><strong>gene universe</strong></span>: <span class="math inline">\({a_{11} + a_{12} + a_{21} + a_{22}} \choose a_{11} + a_{12}\)</span></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/32_choose_15.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="451"></p>
</figure>
</div></li>
<li><p>The ways of selecting <span class="math inline">\(a_{11}\)</span> = 9 genes (number of <span style="color: #00C5CD"><strong>BP+</strong></span> <span style="color: #EE3A8C"><strong>DEGs</strong></span>) without replacement out of the <span class="math inline">\(a_{11} + a_{21}\)</span> = 12 genes in the functional set (<span style="color: #00C5CD"><strong>BP+</strong></span> genes): <span class="math inline">\({{a_{11} + a_{21}} \choose a_{11}}\)</span></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/12_choose_9.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="351"></p>
</figure>
</div></li>
<li><p>The ways of selecting <span class="math inline">\(a_{12}\)</span> = 6 genes (number of <span style="color: #C1CDCD"><strong>BP-</strong></span> <span style="color: #EE3A8C"><strong>DEGs</strong></span>) without replacement out of the <span class="math inline">\(a_{12}+a_{22}\)</span> = 20 genes not in the functional set (<span style="color: #C1CDCD"><strong>BP-</strong></span> genes): <span class="math inline">\({a_{12}+a_{22}}\choose{a_{12}}\)</span></p></li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/20_choose_6.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="355"></p>
</figure>
</div></li>
</ol>
<p>With these 3 values we define the probability of each table arrangement by:</p>
<p><span class="math display">\[
p=\frac{{{a_{11}+a_{21}}\choose a_{11}} \times {{a_{12}+a_{22}}\choose{a_{12}}} }{{a_{11}+a_{12}+a_{21}+a_{22}}\choose{a_{11}+a_{12}}}
\]</span></p>
<p>This is precisely the probability function for the <a href="#0">hypergeometric distribution</a>, used for sampling without replacement and which we use to test the null hypothesis.</p>
<p><span class="math display">\[
p(x)=\frac{{m \choose x} {n \choose {k-x}}}{{m+n}\choose k}\]</span></p>
<ul>
<li><p><span class="math inline">\(k\)</span> is the sample size (number of <span style="color: #EE3A8C"><strong>DEGs</strong></span>) = <span class="math inline">\(a_{11} + a_{12}\)</span> = 15</p></li>
<li><p><span class="math inline">\(m\)</span> is the number of successes in the population (number of <span style="color: #00C5CD"><strong>BP+</strong></span> genes) = <span class="math inline">\(a_{11} + a_{21}\)</span> = 12</p></li>
<li><p><span class="math inline">\(n\)</span> is the number of failures in the population (number of <span style="color: #C1CDCD"><strong>BP-</strong></span> genes) = <span class="math inline">\(a_{12} + a_{22}\)</span> = 20</p></li>
<li><p><span class="math inline">\(x\)</span> is the number of successes in the sample (number of <span style="color: #00C5CD"><strong>BP+</strong></span> among <span style="color: #EE3A8C"><strong>DEGs</strong></span>) = <span class="math inline">\(a_{11}\)</span> = 9</p></li>
<li><p><span class="math inline">\(k-x\)</span> is the number of failures in the sample (number of <span style="color: #C1CDCD"><strong>BP-</strong></span> among <span style="color: #EE3A8C"><strong>DEGs</strong></span>) = <span class="math inline">\(a_{12}\)</span> = 6</p></li>
<li><p><span class="math inline">\(m+n\)</span> is the population size (size of <span style="color: #595959"><strong>gene universe</strong></span>) = <span class="math inline">\(a_{11} + a_{12} + a_{21} + a_{22}\)</span> = 32</p></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How to interpret the above formula?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Basically, we are quantifying the number of ways of obtaining a sample of size <span class="math inline">\(k\)</span> composed of <span class="math inline">\(x\)</span> successes from a total of <span class="math inline">\(m\)</span> successes in the population (<span class="math inline">\(m+n\)</span>) AND the remaining <span class="math inline">\(k-x\)</span> elements as failures from a total of <span class="math inline">\(n\)</span> failures in the population; of all possible ways of randomly selecting <span class="math inline">\(k\)</span> elements from the population.</p>
</div>
</div>
<p>Transferred to our scenario, think of it as having 15 chances to select genes without replacement from the universe that contains both <span style="color: #00C5CD"><strong>BP+</strong></span> and <span style="color: #C1CDCD"><strong>BP-</strong></span> genes. Repeat the experiment multiple times and count the times that 9 of the 15 genes you selected were <span style="color: #00C5CD"><strong>BP+</strong></span> (out of 12 BP+ genes in total) and the remaining 6 genes <span style="color: #C1CDCD"><strong>BP-</strong></span> (out of 20 BP- genes in total), over the number of experiments; that approximates your probability. This probability describes how likely it is to observe your table numbers (<strong>9</strong> <span style="color: #00C5CD"><strong>BP+</strong></span> and <strong>6</strong> <span style="color: #C1CDCD"><strong>BP-</strong></span> genes in the sample of <strong>15</strong> <span style="color: #EE3A8C"><strong>DEGs</strong></span>) just by chance.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/experiment.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="547"></p>
</figure>
</div>
<p>For our actual contingency table <span class="math inline">\(p(x=9)=\frac{{12 \choose 9} \times {20\choose6}}{32\choose15}\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Calculate manually</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">choose</span>(<span class="dv">12</span>,<span class="dv">9</span>)<span class="sc">*</span><span class="fu">choose</span>(<span class="dv">20</span>,<span class="dv">6</span>) <span class="sc">/</span> <span class="fu">choose</span>(<span class="dv">32</span>,<span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.01507311</code></pre>
</div>
</div>
<p>Let’s use <code>dhyper()</code> to check the density function for the hypergeometric distribution with the above parameters and confirm we obtain the same probability for <span class="math inline">\(x\)</span> = 9.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Hypergeometric distribution parameters:</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="do">##  -x = 0:12 (a11 for all possible tables)</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="do">##  -m = 12 </span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="do">##  -n = 20</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="do">##  -k = 15</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>probabilities <span class="ot">&lt;-</span> <span class="fu">dhyper</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">12</span>), <span class="at">m=</span><span class="dv">12</span>, <span class="at">n=</span><span class="dv">20</span>, <span class="at">k=</span><span class="dv">15</span>, <span class="at">log =</span> <span class="cn">FALSE</span>)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>xs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">12</span>)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x=</span>xs, <span class="at">p=</span>probabilities)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>df </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    x            p
1   0 2.740565e-05
2   1 8.221696e-04
3   2 9.043865e-03
4   3 4.898760e-02
5   4 1.469628e-01
6   5 2.586545e-01
7   6 2.743306e-01
8   7 1.763554e-01
9   8 6.782899e-02
10  9 1.507311e-02
11 10 1.808773e-03
12 11 1.027712e-04
13 12 2.015121e-06</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bar plot</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>p, <span class="at">label=</span>p)) <span class="sc">+</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>     <span class="fu">theme_classic</span>() <span class="sc">+</span> </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill=</span> <span class="st">"beige"</span>, <span class="at">colour=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>     <span class="fu">geom_text</span>(<span class="at">label=</span><span class="fu">signif</span>(probabilities, <span class="dv">2</span>), </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>                             <span class="at">y=</span>probabilities, <span class="at">size=</span><span class="dv">3</span>, <span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">vjust=</span><span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>     <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of BP+ DEGs"</span>, <span class="at">y =</span> <span class="st">"Probability"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="one-sided-test" class="level3">
<h3 class="anchored" data-anchor-id="one-sided-test"><strong>One-sided test</strong></h3>
<p>As mentioned before, we are not only interested in the probability of the observed contingency table but also in those for more extreme contingency tables, i.e., tables with greater <span class="math inline">\(a_{11}\)</span>’s.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/all_tables_extreme.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="494"></p>
</figure>
</div>
<p>We add the probability of the actual table + the probabilities for such more extreme tables in order to get the probability of observing 9 or more <span style="color: #EE3A8C"><strong>DEGs</strong></span> that are <span style="color: #00C5CD"><strong>BP+</strong></span> under the null hypothesis:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="do">## p(x&gt;=9)</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(probabilities[<span class="dv">10</span><span class="sc">:</span><span class="dv">13</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.01698667</code></pre>
</div>
</div>
<p>Since <em>p</em>&lt;0.05 we reject the null hypothesis and accept that there’s an enrichment of <span style="color: #00C5CD"><strong>BP+</strong></span> genes amongst our <span style="color: #EE3A8C"><strong>DEGs</strong></span>, and hence the biological process has a significant positive association with differential gene expression.</p>
<p>This is called a <strong>one-sided Fisher’s exact test</strong> as we are evaluating probabilities of extreme values at only one side of the density curve (values <span class="math inline">\(a_{11}\)</span>≥9 for enrichment; see plot below).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Color in red those x&gt;=9</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">'beige'</span>, <span class="dv">9</span>), <span class="fu">rep</span>(<span class="st">'red'</span>, <span class="dv">4</span>))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>colors <span class="ot">&lt;-</span> colors</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>p, <span class="at">label=</span>p, <span class="at">fill=</span>colors)) <span class="sc">+</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill=</span> colors, <span class="at">colour=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_text</span>(<span class="at">label=</span><span class="fu">signif</span>(probabilities, <span class="dv">2</span>),</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y=</span>probabilities, <span class="at">size=</span><span class="dv">3</span>, <span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">vjust=</span><span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>                <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of BP+ DEGs"</span>, <span class="at">y =</span> <span class="st">"Probability"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We obtain the same result with <code>fisher.test(alternative = ”greater”)</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher.test</span>(m, <span class="at">alternative =</span> <span class="st">"greater"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  m
p-value = 0.01699
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 1.402791      Inf
sample estimates:
odds ratio 
  6.529133 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that the results of this test are the same for the transposed matrix, so if we put one or the other variable in columns or rows doesn’t affect as long as the first element of the table is the number of successes in the sample (BP+ DEGs).&nbsp;</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Matrix</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>m</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    9    6
[2,]    3   14</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Transposed matrix</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">t</span>(m)     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    9    3
[2,]    6   14</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Fisher test on t(m)</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher.test</span>(<span class="fu">t</span>(m), <span class="at">alternative =</span> <span class="st">"greater"</span>)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  t(m)
p-value = 0.01699
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 1.402791      Inf
sample estimates:
odds ratio 
  6.529133 </code></pre>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="assessing-gene-set-depletion" class="level2">
<h2 class="anchored" data-anchor-id="assessing-gene-set-depletion">Assessing gene-set depletion?</h2>
<p>What if we are not only interested in assessing the over-representation of gene sets in our experimentally-derived gene list, but also in their under-representation? There may be fewer <span style="color: #00C5CD"><strong>BP+</strong></span> genes in the group of <span style="color: #EE3A8C"><strong>DEGs</strong></span> than expected by chance. In such case we’d be assessing for a negative association between the variables and we need to sum the probabilities of our table and those that are less extreme.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Color in red x&lt;=9</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">'red'</span>, <span class="dv">10</span>), <span class="fu">rep</span>(<span class="st">'beige'</span>, <span class="dv">3</span>))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>colors <span class="ot">&lt;-</span> colors</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>p, <span class="at">label=</span>p, <span class="at">fill=</span>colors)) <span class="sc">+</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill=</span> colors, <span class="at">colour=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_text</span>(<span class="at">label=</span><span class="fu">signif</span>(probabilities, <span class="dv">2</span>),</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y=</span>probabilities, <span class="at">size=</span><span class="dv">3</span>, <span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">vjust=</span><span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>                <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of BP+ DEGs"</span>, <span class="at">y =</span> <span class="st">"Probability"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In this case we use <code>fisher.test(alternative = ”less”)</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="do">## p-value:</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(probabilities[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.9980864</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Same with fisher.test(alternative = "less")</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher.test</span>(m, <span class="at">alternative =</span> <span class="st">"less"</span>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  m
p-value = 0.9981
alternative hypothesis: true odds ratio is less than 1
95 percent confidence interval:
  0.00000 37.74069
sample estimates:
odds ratio 
  6.529133 </code></pre>
</div>
</div>
</section>
<section id="two-sided-test" class="level2">
<h2 class="anchored" data-anchor-id="two-sided-test"><strong>Two-sided test</strong></h2>
<p>The <strong>two-sided Fisher’s exact test</strong> is based on both tails of the hypergeometric distribution and thus is used when we are interested in both enrichment and depletion, i.e., if there’s any association between the variables irrespective of the sign.</p>
<p>One approach to account for extreme values on both sides is to double the sum of the extreme tables’ probabilities at one side (one-sided <em>p</em>-value):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="do">## p-value</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span><span class="sc">*</span><span class="fu">sum</span>(probabilities[<span class="dv">10</span><span class="sc">:</span><span class="dv">13</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.03397334</code></pre>
</div>
</div>
<p>A second approach is to add the one-sided <em>p</em>-value + all the probabilities that are less than or equal to the one for the observed table (probabilities for <span class="math inline">\(x\)</span> = 0,1, and 2 in this example; see plot below).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="do">## p-value</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(probabilities[<span class="dv">10</span><span class="sc">:</span><span class="dv">13</span>]) <span class="sc">+</span> <span class="fu">sum</span>(probabilities[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.02688011</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Two-sided test</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">'red'</span>, <span class="st">'beige'</span>, <span class="st">'red'</span>), <span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">6</span>,<span class="dv">4</span>))</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>df<span class="sc">$</span>colors <span class="ot">&lt;-</span> colors</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x=</span>x, <span class="at">y=</span>p, <span class="at">label=</span>p, <span class="at">fill=</span>colors)) <span class="sc">+</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">fill=</span> colors, <span class="at">colour=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_text</span>(<span class="at">label=</span><span class="fu">signif</span>(probabilities, <span class="dv">2</span>),</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">y=</span>probabilities, <span class="at">size=</span><span class="dv">3</span>, <span class="at">hjust=</span><span class="fl">0.5</span>, <span class="at">vjust=</span><span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>                <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of BP+ DEGs"</span>, <span class="at">y =</span> <span class="st">"Probability"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We get the same result using <code>fisher.test(alternative = ”two.sided”)</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="do">## With fisher.test(alternative = ”two.sided”)</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fisher.test</span>(m, <span class="at">alternative =</span> <span class="st">"two.sided"</span>)   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  m
p-value = 0.02688
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
  1.120742 51.566943
sample estimates:
odds ratio 
  6.529133 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What’s the odds ratio?
</div>
</div>
<div class="callout-body-container callout-body">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/odds_m.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin: 0px" width="334"></p>
</figure>
</div>
<ul>
<li><p>Odds of <strong><span style="color: #EE3A8C">DE</span></strong> in <strong><span style="color: #00C5CD">BP+</span></strong> genes = <span class="math inline">\(\frac{a_{11}}{a_{21}}\)</span>: how many times it is more likely to be a <span style="color: #EE3A8C"><strong>DEG</strong></span> than a <span style="color: #D8BFD8"><strong>non-DEG</strong></span> among the genes in the functional set.</p></li>
<li><p>Odds of <span style="color: #EE3A8C"><strong>DE</strong></span> in <span style="color: #C1CDCD"><strong>BP-</strong></span> genes =<span class="math inline">\(\frac{a_{12}}{a_{22}}\)</span>: how many times it is more likely to be a <span style="color: #EE3A8C"><strong>DEG</strong></span> than a <span style="color: #D8BFD8"><strong>non-DEG</strong></span> among the genes outside the functional set.</p></li>
</ul>
<p>Then,</p>
<ul>
<li><strong>Odds of <span style="color: #EE3A8C">DE</span> in <span style="color: #00C5CD">BP+</span> vs.&nbsp;<span style="color: #C1CDCD">BP-</span> genes</strong> = <span class="math inline">\(\frac{a_{11}}{a_{21}} / \frac{a_{12}}{a_{22}}\)</span> = <span class="math inline">\(\frac{a_{11}\times a_{22}}{a_{12} \times a_{21}}\)</span>: how many times it is more probable to be a <span style="color: #EE3A8C"><strong>DEG</strong></span> among the <span style="color: #00C5CD"><strong>BP+</strong></span> genes than among <span style="color: #C1CDCD"><strong>BP-</strong></span> genes.</li>
</ul>
<p>Same thing if we interchange the columns and rows:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/odds_tm.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin: 0px" width="331"></p>
</figure>
</div>
<ul>
<li><p>Odds of being <strong><span style="color: #00C5CD">BP+</span></strong> in <strong><span style="color: #EE3A8C">DEGs</span></strong> = <span class="math inline">\(\frac{a_{11}}{a_{12}}\)</span></p></li>
<li><p>Odds of being <strong><span style="color: #00C5CD">BP+</span></strong> in <strong><span style="color: #D8BFD8">non-DEGs</span></strong> = <span class="math inline">\(\frac{a_{21}}{a_{22}}\)</span></p></li>
<li><p><strong>Odds ratio (OR) =</strong> odds of being <span style="color: #00C5CD"><strong>BP+</strong></span> in <span style="color: #EE3A8C"><strong>DEGs</strong></span> vs.&nbsp;<strong><span style="color: #D8BFD8">non-DEGs</span></strong>: <span class="math inline">\(\frac{a_{11}}{a_{12}} / \frac{a_{21}}{a_{22}}\)</span> =<span class="math inline">\(\frac{a_{11}\times a_{22}}{a_{12} \times a_{21}}\)</span></p></li>
</ul>
<p>If OR&gt;1 then it is more likely to be a <span style="color: #00C5CD"><strong>BP+</strong></span> gene among the group of <span style="color: #EE3A8C"><strong>DEGs</strong></span>→ <strong>positive association</strong> between the two variables.</p>
<p>If OR&lt;1 then is more likely to be a <span style="color: #00C5CD"><strong>BP+</strong></span> gene among the group of <span style="color: #D8BFD8"><strong>non-DEGs</strong></span>→ <strong>negative association</strong> between the two variables.</p>
<p>If OR≠1 then is either more or less likely to be a <span style="color: #00C5CD"><strong>BP+</strong></span> gene among the group of <span style="color: #EE3A8C"><strong>DEGs</strong></span> → there is an <strong>association</strong> between the two variables.</p>
<p>The OR returned by <code>fisher.test()</code> differs slightly from this formula as they are calculated based on the conditional maximum-likelihood estimate.</p>
</div>
</div>
</section>
</section>
<section id="conclusion" class="level1">
<h1>Conclusion</h1>
<p>The one-sided Fisher’s exact test is a broadly used test in genomics to assess the enrichment of gene sets of interest in a list of genes resulting from high-throughput experimental measurements, drawing potential biological implications of interest from such assays. Its simple input requirements and statistical basis make it user-friendly and straightforward to implement. However, there are clarifications and caveats to consider:</p>
<ol type="1">
<li><p>Even though here we focused on the Fisher’s exact test, there are plenty of other methodological and statistical strategies to assess functional enrichment and gene-set enrichment more globally [<a href="https://academic.oup.com/bib/article/22/1/545/5722384" class="uri">https://academic.oup.com/bib/article/22/1/545/5722384</a>]. (IS IT FUNCTIONAL SPECIFICALLY OR GSEA)</p></li>
<li><p>Here we exemplified the performance of this test with DEGs and functional gene sets, but this over-representation approach doesn’t restrict the analysis to them, not even to genes. You can conceive the enormous applicability of this test for any overlapping sets of biological features. The design and interpretation of your analysis will vary from case to case tho.</p></li>
<li><p>Intrinsic limitations of this test have been noted:</p>
<p>i) Even if it can be used to analyze any sets of features as mentioned previously, this test does restrict the analysis to overlaps between feature IDs (in which its simplicity lies), which means that relevant and informative feature metrics are cannot be incorporated, compared to functional class scoring (FCS) methods such as the gene set enrichment analysis (GSEA) developed in <span class="citation" data-cites="subramanian2005">(<a href="#ref-subramanian2005" role="doc-biblioref">Subramanian et al. 2005</a>)</span>.</p>
<p>ii) When analyzing genes, which is usually the case, we assume they are independent features but that’s not necessarily true. Fisher’s test doesn’t take into account gene correlations.</p></li>
<li><p>Last reminder, Fisher’s exact test works to test enrichment but also depletion, careful with what you are looking to assess, how you run this analysis and how you interpret your results!</p></li>
</ol>
<p>Hopefully at this point you have a clearer understanding around this test and feel more prepared to decide if this is appropriate to test your hypothesis based on your data.</p>
</section>
<section id="relevant-links" class="level1">
<h1>Relevant links</h1>
<ol type="1">
<li><p>Functional enrichment analysis of omics data</p>
<p>The Bioconductor 2018 Workshop Compilation, Chapter 9 <a href="https://bioconductor.github.io/BiocWorkshops" class="uri">https://bioconductor.github.io/BiocWorkshops</a></p></li>
<li><p>Material from Bioc2020 (4th edition): <a href="https://github.com/waldronlab/enrichOmics" class="uri">https://github.com/waldronlab/enrichOmics</a> <a href="https://www.youtube.com/watch?v=n0vVwuQSZkg" class="uri">https://www.youtube.com/watch?v=n0vVwuQSZkg</a></p></li>
</ol>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-ashburner2000" class="csl-entry" role="listitem">
Ashburner, Michael, Catherine A. Ball, Judith A. Blake, David Botstein, Heather Butler, J. Michael Cherry, Allan P. Davis, et al. 2000. <span>“Gene Ontology: Tool for the Unification of Biology.”</span> <em>Nature Genetics</em> 25 (1): 25–29. <a href="https://doi.org/10.1038/75556">https://doi.org/10.1038/75556</a>.
</div>
<div id="ref-kanehisa2000" class="csl-entry" role="listitem">
Kanehisa, M. 2000. <span>“KEGG: Kyoto Encyclopedia of Genes and Genomes.”</span> <em>Nucleic Acids Research</em> 28 (1): 27–30. <a href="https://doi.org/10.1093/nar/28.1.27">https://doi.org/10.1093/nar/28.1.27</a>.
</div>
<div id="ref-subramanian2005" class="csl-entry" role="listitem">
Subramanian, Aravind, Pablo Tamayo, Vamsi K. Mootha, Sayan Mukherjee, Benjamin L. Ebert, Michael A. Gillette, Amanda Paulovich, et al. 2005. <span>“Gene Set Enrichment Analysis: A Knowledge-Based Approach for Interpreting Genome-Wide Expression Profiles.”</span> <em>Proceedings of the National Academy of Sciences</em> 102 (43): 15545–50. <a href="https://doi.org/10.1073/pnas.0506580102">https://doi.org/10.1073/pnas.0506580102</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/daianna21\.github\.io\/daianna_blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="lahuuki/lahuuki.github.io" data-repo-id="MDEwOlJlcG9zaXRvcnkzNDM1NTI0NjA=" data-category="General" data-category-id="DIC_kwDOFHoxzM4CfkDR" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
<input type="hidden" id="giscus-base-theme" value="light">
<input type="hidden" id="giscus-alt-theme" value="dark">
</div> <!-- /content -->




</body></html>