<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daianna Gonzalez-Padilla">
<meta name="dcterms.date" content="2024-07-16">
<meta name="description" content="Global overview of MAGMA Gene and Gene-Set analyses and detailed practical demonstration of the SNP-wise gene analysis and the competitive gene-set analysis.">

<title>Daianna Gonzalez-Padilla - MAGMA: How does the Gene and the Gene-Set analysis operate?</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/logo-removebg.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Daianna Gonzalez-Padilla</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">About me</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../materials.qmd"> 
<span class="menu-text">Other materials</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://daianna21.github.io"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="http://www.linkedin.com/in/daianna-glez"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://x.com/daianna_glez/"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:glezdaianna@gmail.com"> <i class="bi bi-envelope-at" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../blog.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#what-youll-learn-here" id="toc-what-youll-learn-here" class="nav-link" data-scroll-target="#what-youll-learn-here">What you’ll learn here</a></li>
  <li><a href="#how-does-it-work" id="toc-how-does-it-work" class="nav-link" data-scroll-target="#how-does-it-work">How does it work?</a>
  <ul class="collapse">
  <li><a href="#gene-analysis" id="toc-gene-analysis" class="nav-link" data-scroll-target="#gene-analysis"><strong>Gene analysis</strong></a>
  <ul class="collapse">
  <li><a href="#multiple-linear-pc-regression-model" id="toc-multiple-linear-pc-regression-model" class="nav-link" data-scroll-target="#multiple-linear-pc-regression-model"><span style="color:#FF7256"><strong>Multiple linear PC regression model</strong></span></a></li>
  <li><a href="#snp-wise-gene-analysis" id="toc-snp-wise-gene-analysis" class="nav-link" data-scroll-target="#snp-wise-gene-analysis"><span style="color:#7AC5CD"><strong>SNP-wise gene analysis</strong></span></a></li>
  </ul></li>
  <li><a href="#gene-set-analysis" id="toc-gene-set-analysis" class="nav-link" data-scroll-target="#gene-set-analysis"><strong>Gene-set analysis</strong></a>
  <ul class="collapse">
  <li><a href="#self-contained-analysis" id="toc-self-contained-analysis" class="nav-link" data-scroll-target="#self-contained-analysis"><span style="color:#8FBC8F"><strong>Self-contained analysis</strong></span></a></li>
  <li><a href="#competitive-analysis" id="toc-competitive-analysis" class="nav-link" data-scroll-target="#competitive-analysis"><span style="color:#CD9B9B"><strong>Competitive analysis</strong></span></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#example-of-snp-wise-gene-analysis-and-competitive-gene-set-analysis" id="toc-example-of-snp-wise-gene-analysis-and-competitive-gene-set-analysis" class="nav-link" data-scroll-target="#example-of-snp-wise-gene-analysis-and-competitive-gene-set-analysis">Example of SNP-wise gene analysis and competitive gene-set analysis</a>
  <ul class="collapse">
  <li><a href="#step-1-annotation-of-snps-onto-genes" id="toc-step-1-annotation-of-snps-onto-genes" class="nav-link" data-scroll-target="#step-1-annotation-of-snps-onto-genes">Step 1: Annotation of SNPs onto genes</a></li>
  <li><a href="#step-2-snp-wise-gene-analysis" id="toc-step-2-snp-wise-gene-analysis" class="nav-link" data-scroll-target="#step-2-snp-wise-gene-analysis">Step 2: SNP-wise gene analysis</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#relevant-links" id="toc-relevant-links" class="nav-link" data-scroll-target="#relevant-links">Relevant links</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">MAGMA: How does the Gene and the Gene-Set analysis operate?</h1>
  <div class="quarto-categories">
    <div class="quarto-category">GWAS</div>
    <div class="quarto-category">Gene Analysis</div>
    <div class="quarto-category">Gene Set Analysis</div>
    <div class="quarto-category">SNP summary statistics</div>
  </div>
  </div>

<div>
  <div class="description">
    <p>Global overview of MAGMA Gene and Gene-Set analyses and detailed practical demonstration of the SNP-wise gene analysis and the competitive gene-set analysis.</p>
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Daianna Gonzalez-Padilla </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 16, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>MAGMA&nbsp;stands for&nbsp;<strong>Multi-marker Analysis of GenoMic Annotation</strong>&nbsp;and is a computational tool for the analysis of the joint effect of multiple genetic markers on a phenotype based on Genome-Wide Association Study (GWAS) data. MAGMA constitutes a flexible and generalizable approach to assess the&nbsp;<strong>association</strong>&nbsp;of&nbsp;<u><strong>individual genes</strong></u>&nbsp;or&nbsp;<u><strong>gene-sets</strong></u>&nbsp;with a&nbsp;<strong>phenotype</strong>&nbsp;based on the associations that the&nbsp;<u><strong>SNPs</strong></u>&nbsp;of the genes have with the same phenotype.</p>
<p>This is a commonly used method that you’ll likely find implemented while reading about GWASes or that you’ll need to run if working with genetic variant associations and gene/gene-set analysis, so let’s see how it works.</p>
</section>
<section id="what-youll-learn-here" class="level1">
<h1>What you’ll learn here</h1>
<ol type="1">
<li><p>Global description of the analyses MAGMA allows to run at gene and gene-set levels.</p></li>
<li><p>A more detailed demonstration of the SNP-wise gene analysis implemented in MAGMA: inputs, commands and outputs ….</p></li>
<li><p>TODO</p></li>
</ol>
</section>
<section id="how-does-it-work" class="level1">
<h1>How does it work?</h1>
<p>MAGMA allows to first perform a gene analysis and subsequently a gene-set analysis based on gene-level results. Note, however, that the gene-set analysis comes as a separate layer and it is an additional optional step, though it is frequently performed and quite useful.</p>
<section id="gene-analysis" class="level2">
<h2 class="anchored" data-anchor-id="gene-analysis"><strong>Gene analysis</strong></h2>
<p>Genetic markers are aggregated to the level of whole individual genes and those in a given gene are jointly tested for their association with the phenotype, defining in such way the association of the gene itself. This definitely allows to detect significant associations of multiple small-effect markers in a gene that individually would be missed, i.e., associations that depend on multiple markers. This also reduces the number of required tests and makes the results more interpretable by summarizing at the gene level.</p>
<p>Two alternatives for gene analysis are offered by MAGMA:</p>
<section id="multiple-linear-pc-regression-model" class="level3">
<h3 class="anchored" data-anchor-id="multiple-linear-pc-regression-model"><span style="color:#FF7256"><strong>Multiple linear PC regression model</strong></span></h3>
<p>If the input data is <strong>raw genotype data</strong>, MAGMA fits a multiple linear PC regression model to the individuals’ phenotypes. Here, for one gene <span class="math inline">\(g\)</span> at a time, the SNP matrix that contains the information of the SNPs for that gene across all the <span class="math inline">\(N\)</span> individuals studied, is projected onto its Principal Components (PCs). These PCs capture genotypic differences between individuals for the given gene and the first <span class="math inline">\(K\)</span> most explanatory ones (comprising 99.9% of the variance) are used as predictors of the phenotype (<span class="math inline">\(Y\)</span>) in the linear model.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/MAGMA_PC_model.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin:0px" width="700"></p>
</figure>
</div>
<p>Phenotype is modeled as:</p>
<p><span class="math display">\[
Y = \alpha_{0g}+X^*_{g}\alpha_g+W\beta_g+\epsilon_g
\]</span></p>
<p><span class="math display">\[
{\begin{bmatrix}Y_1 \\ ... \\ Y_{N-1} \\ Y_N\end{bmatrix}} = {\begin{bmatrix}\alpha_{0g} \\ ... \\ \alpha_{0g} \\ \alpha_{0g}\end{bmatrix}} + \stackrel{PC1 \ \  \ \  \ \ PC2 \ \  \ \  \ \ ...  \ \ \ \ \ \ PC_K} {\begin{bmatrix}  x_{11} \ \ x_{12} \ \ ... \ \ x_{1K}  \\  ... \\ x_{(N-1)1} \ \ x_{(N-1)2} \ \ ... \ \ x_{(N-1)K} \\  x_{N1} \ \ x_{N2} \ \ ... \ \ x_{NK} \end{bmatrix}} {\begin{bmatrix}\alpha_{g1} \\  \alpha_{g2} \\ ... \\ \alpha_{gK}\end{bmatrix}} +  \\ \stackrel{Cov1 \ \  \ \  \ \ Cov2 \ \  \ \  \ \ ...  \ \ \ \ \ \ CovL}{\begin{bmatrix}  w_{11} \ \ w_{12} \ \ ... \ \ w_{1L}  \\  ... \\ w_{(N-1)1} \ \ w_{(N-1)2} \ \ ... \ \ w_{(N-1)L} \\  w_{N1} \ \ w_{N2} \ \ ... \ \ w_{NL} \end{bmatrix}}{\begin{bmatrix}\beta_{g1} \\  \beta_{g2} \\ ... \\ \beta_{gL}\end{bmatrix}} +
{\begin{bmatrix}\epsilon_{g1} \\ ... \\ \epsilon_{g(N-1)} \\  \epsilon_{gN}\end{bmatrix}}
\]</span></p>
<p>With <span class="math inline">\(X_{g}^*\)</span> the matrix of the first <span class="math inline">\(K\)</span> PCs for gene <span class="math inline">\(g\)</span> and <span class="math inline">\(W\)</span> a matrix of additional covariates for the individuals that can be included in the model. <span class="math inline">\(\alpha_g\)</span>’s and <span class="math inline">\(\beta_g\)</span>’s are the coefficients of the PCs and covariates for the gene <span class="math inline">\(g\)</span>, respectively. <span class="math inline">\(\epsilon_g\)</span> is the vector of residuals and <span class="math inline">\(\alpha_{0g}\)</span> the intercept. The <span class="math inline">\(\alpha_g\)</span>’s represent the effect of the genotype for gene <span class="math inline">\(g\)</span> on the phenotype (the <strong>genetic effect</strong>).</p>
<p>Then an F-test is used to assess if genotypes for gene <span class="math inline">\(g\)</span> have an effect on the phenotype under the null hypothesis that the genetic effect of <span class="math inline">\(g\)</span> on the phenotype is 0 across all PCs (<span class="math inline">\(H_0:\alpha_g=\overrightarrow 0\)</span>). This leads to a <em>p</em>-value for the association of gene <span class="math inline">\(g\)</span> with the phenotype <span class="math inline">\(Y\)</span> based on its genetic markers.</p>
<div style="background-color:#F2F2F2; padding: 15px">
<p>✔️ The advantage of using PCs instead of variables for individual SNPs is that 1) the number of variables included in the linear model is reduced and 2) they allow to account for redundancy and collinearity between SNPs. Also this model offers flexibility to accommodate additional covariates to model the phenotype.</p>
</div>
</section>
<section id="snp-wise-gene-analysis" class="level3">
<h3 class="anchored" data-anchor-id="snp-wise-gene-analysis"><span style="color:#7AC5CD"><strong>SNP-wise gene analysis</strong></span></h3>
<p>If no raw genotype data is available, MAGMA also accepts <u><strong>SNP <em>p</em>-values</strong></u> from a GWAS as input. In this case SNP <em>p</em>-values for a gene are first transformed into <u><strong>Z or</strong> <span class="math inline">\(\chi^2\)</span><strong>statistics</strong></u> and then combined into a <u><strong>gene test-statistic</strong></u> by the <span style="background-color: #F0FFF0"><strong>mean SNP statistic</strong></span> or <span style="background-color:#FFF0F5"><strong>top SNP statistic</strong></span> method. This gene test-statistic is then used to compute the <u><strong>gene <em>p</em>-value</strong></u>, either by an approximation of the sampling distribution (mean SNP statistic method) or by phenotype permutation (mean and top SNP statistic methods).</p>
<p>This SNP-wise MAGMA analysis based on SNP summary statistics requires a <strong>reference dataset</strong> with similar ancestry as the data from which SNPs <em>p</em>-values were computed. This is needed to account for linkage disequilibrium (LD) between SNPs. If an approximation of the sampling distribution is used to compute the gene <em>p</em>-value, this reference is required to obtain the SNP statistic correlation matrix <span class="math inline">\(R\)</span>. If phenotype permutation is chosen, these reference data are necessary to generate the gene test-statistics empirical sampling distribution.</p>
<div style="background-color:#F2F2F2; padding: 15px">
<p>Summary: SNP <em>p</em>-values → SNP statistics →<span style="color:#EE00EE"><strong>*</strong></span>gene test-statistic →<span style="color: #00CD00"><strong>*</strong></span> gene <em>p</em>-value → gene <em>z</em>-score (<span class="math inline">\(z_g\)</span>)</p>
</div>
<p><span style="color:#EE00EE"><strong>*</strong></span>Gene test-statistics can be constructed from the SNP statistics applying one of the two implemented methods in MAGMA:</p>
<ul>
<li><p><span style="background-color: #F0FFF0"><strong>Mean SNP statistic</strong></span>:</p>
<ul>
<li><p><strong>Mean Z gene test-statistic</strong>: SNP <em>p</em>-values of a gene are transformed to Z-statistics and the mean (weighted or unweighted sum) of such Z-stats is used as the test-statistic of the gene: <span class="math inline">\(T=\sum_i^G w_iz_i\)</span>, with <span class="math inline">\(G\)</span> the total number of SNPs in the gene and <span class="math inline">\(z_i\)</span> and <span class="math inline">\(w_i\)</span> the Z-statistic and weight for SNP <span class="math inline">\(i\)</span>, respectively. These gene test-stats follow a normal distribution.</p></li>
<li><p><strong>Mean</strong> <span class="math inline">\(\chi^2\)</span> <strong>gene test-statistic</strong>: the mean (weighted or unweighted sum) of the <span class="math inline">\(-2log\)</span> of the SNP <em>p</em>-values (that are <span class="math inline">\(\sim\chi^2_{(2)}\)</span>) is used as gene test-statistic: <span class="math inline">\(T=-2\sum_i^G w_i\log(p_i)\)</span>, with <span class="math inline">\(p_i\)</span> the <em>p</em>-value of SNP <span class="math inline">\(i\)</span>. These gene test-stats are assumed to be <span class="math inline">\(\sim c\chi^2_{(f)}\)</span>, where <span class="math inline">\(c\)</span> and <span class="math inline">\(f\)</span> are constants of the distribution.</p></li>
</ul>
<p><span style="color: #00CD00"><strong>*</strong></span>For both mean Z and mean <span class="math inline">\(\chi^2\)</span> gene test-statistics the <em>p</em>-value of the gene is obtained based on a known <strong>approximation of the sampling distribution</strong> of the gene test-stats. For such purposes, the reference dataset is needed to estimate the SNP statistic correlation matrix, from which the optional SNP weights (<span class="math inline">\(w_i\)</span>) are computed to correct for dependency between SNPs. A phenotype permutation procedure can also be applied to compute empirical gene <em>p</em>-values (see further below).</p></li>
<li><p><span style="background-color:#FFF0F5"><strong>Top SNP statistic(s)</strong></span>: the lowest SNP <em>p</em>-value among all SNPs in the gene (corresponding to the most associated SNP) or the sum of the <span class="math inline">\(-log(p)\)</span> for the top most associated SNPs is used as the gene test-statistic.</p>
<p><span style="color: #00CD00"><strong>*</strong></span>To compute the gene <em>p</em>-value a <strong>phenotype permutation</strong> procedure is the only option available. Here random phenotypes are assigned to the reference data and a top gene test-statistic is computed in each permutation. The empirical gene <em>p</em>-value is computed as the proportion of these permuted gene test-statistics that are higher than the observed one.</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The SNP-wise gene analysis of MAGMA can also be run with raw genotype data, in which case SNP <em>p</em>-values are computed internally and the raw genotype data takes place of the reference data.</p>
</div>
</div>
<p>After running the gene analysis, either by PC regression or SNP-wise analysis, the resulting gene <em>p</em>-values are subsequently transformed to <u><strong>Z-values (</strong><span class="math inline">\(z_g\)</span><strong>)</strong></u> to be introduced into the gene-set analysis. These Z-values are normally distributed and capture the <strong>association</strong> <strong>strength</strong> of the genes with the phenotype: the greater (higher positive) they are the stronger the gene association.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that with the gene analysis F-test (with PC regression) and gene test-statistics (with SNP-wise models) we are not really looking at the sign (+/-) of the association of the genetic markers of a gene with the phenotype, just exploring if there’s an association at all:</p>
<ul>
<li><p>In the PC regression model, we only assess if all genetic effects of a gene (<span class="math inline">\(\alpha_g\)</span>’s) are zero or not, but not if the global gene effect is positive or negative.</p></li>
<li><p>With SNP-wise methods, since they are based on SNP <em>p</em>-values that don’t inform about the sign of the association, just how strong it is, the gene test-stats derived from them are also reflecting “sign-ignorant” associations. The gene <em>p</em>-value, however, is computed based on the right tail of the distribution of these gene test-stats (<span class="math inline">\(T\)</span>’s) as they correspond to stronger (but not necessarily positive) associations.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/T_distribution_transp.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin:0px" width="271"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="gene-set-analysis" class="level2">
<h2 class="anchored" data-anchor-id="gene-set-analysis"><strong>Gene-set analysis</strong></h2>
<p>Once having gene-level associations, we can now think of gene set associations. Intiatlly, we aggregate genes in groups of genes that share cellular or functional properties, such as involvements in certain biological processes, with particular molecular functions, cellular locations, cell-type specific expression or activity, or differential expression for a given condition. These define concrete processes or features we may be interested in relating with a GWAS phenotype. By doing this, we can gain insights into the potential molecules, pathways, and tissues implicated in a phenotype’s etiology.</p>
<p>Since all that’s necessary from the gene analysis to run the gene-set analysis are the gene Z-values, gene-set analysis operates the same independently of the gene analysis method applied (PC regression or SNP-wise), as both result in gene <em>p</em>-values and respective <span class="math inline">\(z_g\)</span>’s.</p>
<p>Two gene-level regression models are implemented to asses the association of gene-sets with the phenotype:</p>
<section id="self-contained-analysis" class="level3">
<h3 class="anchored" data-anchor-id="self-contained-analysis"><span style="color:#8FBC8F"><strong>Self-contained analysis</strong></span></h3>
<p>Tests if the genes of a gene set <span class="math inline">\(s\)</span> are jointly associated with the phenotype. This is assessed by an intercept-only linear regression <span class="math inline">\(Z_s=\beta_0\overrightarrow1+\epsilon_s\)</span>, with <span class="math inline">\(Z_s\)</span> the variable with the <span class="math inline">\(z_g\)</span>’s of the genes in <span class="math inline">\(s\)</span>. With this simple model we evaluate the null <span class="math inline">\(H_0:\beta_0=0\)</span> against the alternative <span class="math inline">\(H_A: \beta_0&gt;0\)</span>. This is equivalent to a one-sided single-sample <em>t</em>-test comparing the mean association of genes in <span class="math inline">\(s\)</span> to 0. In other words, we are evaluating if the mean association of the genes in the set is greater than 0 (i.e., is a strong association, not to be confused with a positive effect of <span class="math inline">\(s\)</span> on the phenotype).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GSA_self_contained.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin:0px" width="378"></p>
</figure>
</div>
</section>
<section id="competitive-analysis" class="level3">
<h3 class="anchored" data-anchor-id="competitive-analysis"><span style="color:#CD9B9B"><strong>Competitive analysis</strong></span></h3>
<p>Tests whether the genes in the set <span class="math inline">\(s\)</span> are more strongly associated with the phenotype than the rest of genes not in the set. Here the outcome variable <span class="math inline">\(Z\)</span> of the model now includes the <span class="math inline">\(z_g\)</span> of all genes. We define <span class="math inline">\(S_s\)</span> as a binary predictor of <span class="math inline">\(Z\)</span> that equals 1 if a gene is in <span class="math inline">\(s\)</span> and 0 if not: <span class="math inline">\(Z=\beta_{0s}\overrightarrow1+S_s\beta_s+\epsilon\)</span>.</p>
<p>Here <span class="math inline">\(\beta_s\)</span> reflects the difference in the association strength of the genes in <span class="math inline">\(s\)</span> versus those not in <span class="math inline">\(s\)</span>. The <em>p</em>-value of the gene set results from testing on <span class="math inline">\(\beta_s\)</span>: we test <span class="math inline">\(H_0: \beta_s=0\)</span> vs <span class="math inline">\(H_A: \beta_s&gt;0\)</span>, which is equivalent to a one-sided two-sample <em>t</em>-test evaluating if the mean association strength of genes in <span class="math inline">\(s\)</span> is greater than the mean association strength of genes not in <span class="math inline">\(s\)</span>. But again, this doesn’t mean genes in the set <span class="math inline">\(s\)</span> have a positive association with the phenotype. This analysis is performed by default.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/GSA_competitive.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="margin:0px" width="228"></p>
</figure>
</div>
<div style="background-color:#F2F2F2; padding: 15px">
<p>✔️ Both self-contained and competitive gene-set analyses are implemented through more general gene-level regression models that can accommodate combinations of gene sets and gene properties such as gene expression levels and gene size to test for their effects and correct for potential confounding effects. This can be useful, for example, to test if variables such as differential expression between tissues or conditions have an effect on the phenotype associations.</p>
<p>Further, analyzing gene sets evidently reduces the number of output tests, but note that one test per gene still has to be performed in the the previous gene analysis.</p>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Due to LD between SNPs neighboring genes are usually correlated. This obligate us to assume <span class="math inline">\(\epsilon \sim MVN(\overrightarrow0, \sigma^2R)\)</span> in the gene-set analysis, where <span class="math inline">\(R\)</span> is the <strong>gene-gene correlation matrix</strong>.</p>
</div>
</div>
</section>
</section>
</section>
<section id="example-of-snp-wise-gene-analysis-and-competitive-gene-set-analysis" class="level1">
<h1>Example of SNP-wise gene analysis and competitive gene-set analysis</h1>
<p>By default the SNP-wise gene analysis method implemented is the mean Z gene test-statistic, and the default gene-set analysis is the competitive one. Let’s exemplify how to run these analyses. The following are the three main steps implemented by MAGMA.</p>
<section id="step-1-annotation-of-snps-onto-genes" class="level2">
<h2 class="anchored" data-anchor-id="step-1-annotation-of-snps-onto-genes">Step 1: Annotation of SNPs onto genes</h2>
<p>The first step is to map all SNPs from your GWAS data onto the genes of a reference genome.</p>
<ul>
<li><p><strong>Input</strong>: we need SNP and gene positions based on the same human genome reference build; we provide this in the&nbsp;<code>.snploc</code>&nbsp;and in the&nbsp;<code>.gene.loc</code>&nbsp;files, respectively.</p>
<div class="sourceCode" id="cb1" style="padding: 14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">## .snploc file example</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Columns: SNP ID, Chr, Position:</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>       <span class="ex">SNP</span> CHR         BP </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">rs62513865</span>   8  101592213 </span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">rs79643588</span>   8  106973048 </span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">rs17396518</span>   8  108690829 </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="ex">rs983166</span>     8  108681675 </span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">rs28842593</span>   8  103044620 </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">rs7014597</span>    8  104152280 </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">rs3134156</span>    8  100479917 </span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ex">rs6980591</span>    8  103144592 </span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="ex">rs72670434</span>   8  108166508</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb2" style="padding: 14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">## .gene.loc file example</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Required columns: GeneID, Chr, Start site, Stop site:</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000243485</span> 1 29554 31109 + MIR1302-2HG </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000186092</span> 1 65419 71585 + OR4F5 </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000238009</span> 1 89295 133723 <span class="at">-</span> AL627309.1 </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000239906</span> 1 139790 140339 <span class="at">-</span> AL627309.2 </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000236601</span> 5 180881343 180888537 <span class="at">-</span> AL732372.1 </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000235146</span> 1 523009 530148 + AC114498.1 </span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000229905</span> 1 696291 697369 + AL669831.2 </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000237491</span> 1 714150 745440 + AL669831.5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>SNPs in sexual chromosomes are usually excluded from the analysis as their inclusion brings additional analytical, statistical, and bioinformatic challenges for their analysis <span class="citation" data-cites="sun2023">(<a href="#ref-sun2023" role="doc-biblioref">Sun et al. 2023</a>)</span>. MAGMA only supports X chromosome analysis.</p>
</div>
</div></li>
<li><p><strong>Commands</strong>: the following are the flags and inputs used to map the SNPs onto genes.</p>
<div class="sourceCode" id="cb3" style="padding:14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Run SNP annotation: </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">## --annotate: flag to run annotation </span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">## --snp-loc: indicate file with SNP positions</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">## --gene-loc: indicate file with gene positions</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">## --out: prefix for output files</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="ex">magma</span> <span class="at">--annotate</span> <span class="dt">\ </span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>      <span class="ex">--snp-loc</span> [.snploc file]<span class="dt">\ </span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>      <span class="ex">--gene-loc</span> [gene.loc file]<span class="dt">\ </span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>      <span class="ex">--out</span> <span class="pp">[</span><span class="ss">output_prefix</span><span class="pp">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><strong>Output</strong>: this step returns the <code>.genes.annot</code> file as output, containing all SNPs that were assigned to a gene.</p>
<div class="sourceCode" id="cb4" style="padding:14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">## .genes.annot file example</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">## GeneID Chr:Start:Stop and corresponding SNP rsIDs:</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">window_up</span> = 0 </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">window_down</span> = 0 </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000237491</span> 1:714150:745440 rs186002080 rs12184267 rs142557973 rs149887893 </span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000177757</span> 1:752751:755217 rs3115859 rs3131968 rs3131967 rs3115860 </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000225880</span> 1:761586:762902 rs2286139 rs377377186 rs374493323 </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="ex">ENSG00000230368</span> 1:803451:812283 rs11240779 rs28410559 rs72631880 rs58686784 </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="step-2-snp-wise-gene-analysis" class="level2">
<h2 class="anchored" data-anchor-id="step-2-snp-wise-gene-analysis">Step 2: SNP-wise gene analysis</h2>
<ul>
<li><p><strong>Input</strong>: we need to provide the SNP <em>p</em>-values in the <code>.pval</code> file and the gene-wise SNPs in the <code>.genes.annot</code> file obtained in the previous step. A reference dataset must be provided as well.</p>
<p>The total sample size (N) of the GWAS from which the SNP <em>p</em>-values were derived must be indicated as well using the <em>N</em> modifier of <code>--pval</code>. Alternatively, if different sample sizes were used per SNP, these can be contained in a separate column of the <code>.pval</code> file which specified in the <em>ncol</em> modifier of <code>--pval</code>.</p>
<div class="sourceCode" id="cb5" style="padding:14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">## .pval file example</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Columns: SNP ID, SNP p-value, and optionally SNP sample size </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>       <span class="ex">SNP</span>      P      N </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="ex">rs62513865</span> 0.4847 130644 </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="ex">rs79643588</span> 0.5605 130644 </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="ex">rs17396518</span> 0.8145 130644 </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="ex">rs983166</span>   0.5704 130644 </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="ex">rs28842593</span> 0.7488 130644 </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="ex">rs7014597</span>  0.5034 130644 </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="ex">rs3134156</span>  0.2225 130644</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The next is a <code>head</code> of the 1000 Genomes European panel data used as the reference dataset.</p>
<div class="sourceCode" id="cb6" style="padding:14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">## g1000_eur reference data example</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Columns: Chr, SNP ID, Genetic distance, Position, Allele1, Allele2</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> rs537182016 0 10539 A C </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> rs575272151 0 11008 G C </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> rs544419019 0 11012 G C </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> rs540538026 0 13110 A G </span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> rs62635286  0 13116 G T </span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="ex">1</span> rs200579949 0 13118 G A </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><strong>Commands</strong>: these are the flags used to run the SNP-wise mean Z stats method.</p>
<p>## Reference data used was the 1000 Genomes European data</p>
<div class="sourceCode" id="cb7" style="padding:14px"><pre class="sourceCode Bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Run gene analysis based on SNP p-values</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">## --bfile</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">## --pval</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">## --gene-annot</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">## --out</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="ex">magma</span> </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="ex">--bfile</span> g1000_eur<span class="dt">\ </span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">## SNP p-values as input; </span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">## indicate colname for sample size of the study (N) </span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="ex">--pval</span> <span class="va">$ANNOT_PREFIX</span>.pval ncol=N<span class="dt">\ </span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co">## Provide output file from step 1 with gene SNPs </span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="ex">--gene-annot</span> <span class="va">$OUTPUT_PREFIX</span>.genes.annot<span class="dt">\ </span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="ex">--out</span> <span class="va">$OUTPUT_PREFIX</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
</section>
<section id="conclusion" class="level1">
<h1>Conclusion</h1>
<p>MAGMA is a powerful computational tool for extensible gene and gene-set analyses. By developing these two analyses as separate layers,</p>
</section>
<section id="relevant-links" class="level1">
<h1>Relevant links</h1>
<div style="background-color: #F8F8FF;  padding: 14px; margin-left: 0px; margin-bottom: 1px; font-size: 15px; line-height: 1.3">
<div class="line-block">📑 <strong>Original publication for MAGMA:</strong><br>
de Leeuw, C. A., Mooij, J. M., Heskes, T., &amp; Posthuma, D. (2015). MAGMA: generalized gene-set analysis of GWAS data.&nbsp;<em>PLoS computational biology</em>,&nbsp;<em>11</em>(4), e1004219. <a href="https://doi.org/10.1371/journal.pcbi.1004219https://doi.org/10.1371/journal.pcbi.1004219">https://doi.org/10.1371/journal.pcbi.1004219</a></div>
</div>
<p>&nbsp;</p>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>Contents of this post are based on the original MAGMA publication and Supplementary File 1, and the official documentation website linked above.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-sun2023" class="csl-entry" role="listitem">
Sun, Lei, Zhong Wang, Tianyuan Lu, Teri A. Manolio, and Andrew D. Paterson. 2023. <span>“eXclusionarY: Ten Years Later, Where Are the Sex Chromosomes in GWAS?”</span> <a href="http://dx.doi.org/10.1101/2023.02.03.526992">http://dx.doi.org/10.1101/2023.02.03.526992</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/daianna21\.github\.io\/daianna_blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>