<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daianna Gonzalez-Padilla">
<meta name="dcterms.date" content="2024-12-12">

<title>Daianna Gonzalez-Padilla - What comes first: count normalization or gene/sample filtering?</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/logo-removebg.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Daianna Gonzalez-Padilla</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">About me</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://daianna21.github.io"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="http://www.linkedin.com/in/daianna-glez"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://x.com/daianna_glez/"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:glezdaianna@gmail.com"> <i class="bi bi-envelope-at" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../blog.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#what-youll-learn-here" id="toc-what-youll-learn-here" class="nav-link" data-scroll-target="#what-youll-learn-here">What you’ll learn here</a></li>
  <li><a href="#an-example-data-set" id="toc-an-example-data-set" class="nav-link" data-scroll-target="#an-example-data-set">An example data set</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a>
  <ul class="collapse">
  <li><a href="#counts-per-million" id="toc-counts-per-million" class="nav-link" data-scroll-target="#counts-per-million">Counts per million</a></li>
  <li><a href="#tmm" id="toc-tmm" class="nav-link" data-scroll-target="#tmm">TMM</a></li>
  </ul></li>
  <li><a href="#filtering-lowly-expressed-genes" id="toc-filtering-lowly-expressed-genes" class="nav-link" data-scroll-target="#filtering-lowly-expressed-genes">Filtering lowly-expressed genes</a>
  <ul class="collapse">
  <li><a href="#cpm-cuttoff" id="toc-cpm-cuttoff" class="nav-link" data-scroll-target="#cpm-cuttoff">CPM cuttoff</a></li>
  <li><a href="#filterbyexpr" id="toc-filterbyexpr" class="nav-link" data-scroll-target="#filterbyexpr">filterByExpr</a></li>
  </ul></li>
  <li><a href="#filtering-low-quality-samples" id="toc-filtering-low-quality-samples" class="nav-link" data-scroll-target="#filtering-low-quality-samples">Filtering low-quality samples</a>
  <ul class="collapse">
  <li><a href="#qc-metrics-based-on-rawlog-normalized-counts" id="toc-qc-metrics-based-on-rawlog-normalized-counts" class="nav-link" data-scroll-target="#qc-metrics-based-on-rawlog-normalized-counts">QC metrics based on raw/log-normalized counts</a></li>
  <li><a href="#qc-metrics-based-on-allexpressed-genes" id="toc-qc-metrics-based-on-allexpressed-genes" class="nav-link" data-scroll-target="#qc-metrics-based-on-allexpressed-genes">QC metrics based on all/expressed genes</a></li>
  </ul></li>
  <li><a href="#the-order-matters" id="toc-the-order-matters" class="nav-link" data-scroll-target="#the-order-matters">The order matters</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">What comes first: count normalization or gene/sample filtering?</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Normalization</div>
    <div class="quarto-category">Filtering</div>
    <div class="quarto-category">Trimmed Mean of M values</div>
    <div class="quarto-category">Lowly-expressed genes</div>
    <div class="quarto-category">Low-quality samples</div>
    <div class="quarto-category">RNA-seq</div>
  </div>
  </div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Daianna Gonzalez-Padilla </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 12, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>⚠️ This page is under development.</p>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>One of my biggest questions while analyzing RNA-seq data has always been if I should first filter lowly-expressed genes and poor-quality samples and then normalize raw counts, or in what order these steps should be executed. Is there an order to follow irrespective of the filtering and normalization methods used? Does the order actually matter?</p>
<p>As you can imagine, if I’m writting this post is because I haven’t found a clear answer to all these questions. In my search I have gained more doubts than answers, but I have realized this is not an unaddressed issue but a hard one to close. With this post demonstrations I can finally say the order does matter, but it depends on the methodologies applied for filtering and normalizing.</p>
<p>In this post I will show the implications of the ordering of filtering and normalization steps, depending on the methods and metrics used for these analyses, with special attention to the TMM normalization method. I will offer warnings and practical recommendations for you to proceed with your analysis without racking you brain about it.</p>
</section>
<section id="what-youll-learn-here" class="level1">
<h1>What you’ll learn here</h1>
</section>
<section id="an-example-data-set" class="level1">
<h1>An example data set</h1>
<p>Using <a href="https://rna.recount.bio"><em>recount3</em></a> <span class="citation" data-cites="leonardocollado-torres2020 wilks2021a">(<a href="#ref-leonardocollado-torres2020" role="doc-biblioref"><strong>leonardocollado-torres2020?</strong></a>; <a href="#ref-wilks2021a" role="doc-biblioref"><strong>wilks2021a?</strong></a>)</span> we will download data from a study (<a href="10.1016/j.chembiol.2019.05.005">Marc Hafner et al., 2019</a>; SRA ID: <code>SRP107565</code>) where 216 bulk samples were sequenced to examine the transcriptomic changes induced by 3 drugs for treating breast cancer.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(recount3)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Download all available projects in human in recount3</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>human_projects <span class="ot">&lt;-</span> <span class="fu">available_projects</span>()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Download gene expression data and metadata from SRP107565 study</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>proj_info <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    human_projects,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    project <span class="sc">==</span> <span class="st">"SRP107565"</span> <span class="sc">&amp;</span> project_type <span class="sc">==</span> <span class="st">"data_sources"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">recount3_url =</span> <span class="st">"https://sciserver.org/public-data/recount3/data"</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>proj_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    project organism file_source     project_home project_type n_samples
1 SRP107565    human         sra data_sources/sra data_sources       216</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create RangedSummarizedExperiment object to handle RNA-seq and sample data </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>rse <span class="ot">&lt;-</span> <span class="fu">create_rse</span>(proj_info)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Gene expression data in assay(rse): for first 5 genes and 5 samples</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(rse)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  SRR5579425 SRR5579426 SRR5579433 SRR5579434 SRR5579435
ENSG00000278704.1          0          0          0          0          0
ENSG00000277400.1          0          0          0          0          0
ENSG00000274847.1          0          0          0          0          0
ENSG00000277428.1          0          0          0          0          0
ENSG00000276256.1          0          0          0          0          0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Sample metadata in colData(rse): for first 6 samples and 3 variables</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(rse)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
             rail_id external_id       study
           &lt;integer&gt; &lt;character&gt; &lt;character&gt;
SRR5579425   1000031  SRR5579425   SRP107565
SRR5579426   1000045  SRR5579426   SRP107565
SRR5579433   1000255  SRR5579433   SRP107565
SRR5579434   1000270  SRR5579434   SRP107565
SRR5579435   1000286  SRR5579435   SRP107565
SRR5579436   1000301  SRR5579436   SRP107565</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Gene data in rowData(rse): for first 6 genes and 3 variables</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(rse)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
                    source     type bp_length
                  &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
ENSG00000278704.1  ENSEMBL     gene      2237
ENSG00000277400.1  ENSEMBL     gene      2179
ENSG00000274847.1  ENSEMBL     gene      1599
ENSG00000277428.1  ENSEMBL     gene       101
ENSG00000276256.1  ENSEMBL     gene      2195
ENSG00000278198.1  ENSEMBL     gene      1468</code></pre>
</div>
</div>
<p>In the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html"><em>RangedSummarizedExperiment</em></a> <span class="citation" data-cites="martinmorgan2017 huber2015">(<a href="#ref-martinmorgan2017" role="doc-biblioref"><strong>martinmorgan2017?</strong></a>; <a href="#ref-huber2015" role="doc-biblioref"><strong>huber2015?</strong></a>)</span> object (<code>rse</code>) there are raw expression data for 63,856 genes across the 216 samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(rse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 63856   216</code></pre>
</div>
</div>
</section>
<section id="normalization" class="level1">
<h1>Normalization</h1>
<p>In order to compare gene expression measures between samples of two or more experimental groups, normalization of raw counts is required to correct for between-sample technical differences, such as sequencing depth. Various normalization methods have been proposed, each relying on specific methodological assumptions [].</p>
<p>One of the most commonly adopted schemes is the trimmed mean of M values method.</p>
<section id="normalization-methods-on-rna-seq" class="level4">
<h4 class="anchored" data-anchor-id="normalization-methods-on-rna-seq"><strong>Normalization Methods on RNA-Seq:</strong></h4>
</section>
<section id="httpspmc.ncbi.nlm.nih.govarticlespmc4484837" class="level4">
<h4 class="anchored" data-anchor-id="httpspmc.ncbi.nlm.nih.govarticlespmc4484837">- <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4484837/" class="uri">https://pmc.ncbi.nlm.nih.gov/articles/PMC4484837/</a></h4>
<ul>
<li><a href="https://academic.oup.com/bib/article/19/5/776/3056951" class="uri">https://academic.oup.com/bib/article/19/5/776/3056951</a></li>
</ul>
<p>Comparison of normalization methods</p>
<ul>
<li><a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-2353-z" class="uri">https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-2353-z</a></li>
</ul>
</section>
<section id="counts-per-million" class="level2">
<h2 class="anchored" data-anchor-id="counts-per-million">Counts per million</h2>
</section>
<section id="tmm" class="level2">
<h2 class="anchored" data-anchor-id="tmm">TMM</h2>
<p><a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC2864565/pdf/gb-2010-11-3-r25.pdf" class="uri">https://pmc.ncbi.nlm.nih.gov/articles/PMC2864565/pdf/gb-2010-11-3-r25.pdf</a></p>
</section>
</section>
<section id="filtering-lowly-expressed-genes" class="level1">
<h1>Filtering lowly-expressed genes</h1>
<section id="cpm-cuttoff" class="level2">
<h2 class="anchored" data-anchor-id="cpm-cuttoff">CPM cuttoff</h2>
</section>
<section id="filterbyexpr" class="level2">
<h2 class="anchored" data-anchor-id="filterbyexpr">filterByExpr</h2>
</section>
</section>
<section id="filtering-low-quality-samples" class="level1">
<h1>Filtering low-quality samples</h1>
<section id="qc-metrics-based-on-rawlog-normalized-counts" class="level2">
<h2 class="anchored" data-anchor-id="qc-metrics-based-on-rawlog-normalized-counts">QC metrics based on raw/log-normalized counts</h2>
</section>
<section id="qc-metrics-based-on-allexpressed-genes" class="level2">
<h2 class="anchored" data-anchor-id="qc-metrics-based-on-allexpressed-genes">QC metrics based on all/expressed genes</h2>
<p>Samples are required in the filtering process and in TMM (sample vs sample comparison; in CPM library sizes are computed independently)</p>
<p>For sample outlier detection we run PCA with normalized and filtered counts.</p>
<p>It depends on if we have prior knowledge of the samples quality and if those QC metrics that are outliers imply that the RNA composition is atypical or problematic: we can plot that and compare. If there are few samples they are not likely to affect downstream results.</p>
</section>
</section>
<section id="the-order-matters" class="level1">
<h1>The order matters</h1>
<p>What comes first normalization or filtering of genes and samples? Is TMM afected by filtering genes and samples?</p>
<p>Cases:</p>
<ol type="1">
<li>Filtering genes –&gt; filtering samples –&gt; normalization (with librari size only OR TMM)</li>
<li>Filtering genes –&gt; normalization –&gt; filtering samples</li>
<li>Filtering samples –&gt; filtering genes –&gt; normalization</li>
<li>Filtering samples –&gt; normalization –&gt; filtering genes</li>
<li>Normalization –&gt; filtering samples –&gt; filtering genes</li>
<li>Normalization –&gt; filtering genes –&gt; filtering samples</li>
</ol>
<section id="section" class="level4">
<h4 class="anchored" data-anchor-id="section"></h4>
<p>Order or norm and filtering:</p>
<ul>
<li><p><a href="https://support.bioconductor.org/p/116351/#:~:text=If%20you%20must%20filter%20on,same%20procedure%20as%20described%20above." class="uri">https://support.bioconductor.org/p/116351/#:~:text=If%20you%20must%20filter%20on,same%20procedure%20as%20described%20above.</a></p></li>
<li><p><a href="https://support.bioconductor.org/p/123624/" class="uri">https://support.bioconductor.org/p/123624/</a></p></li>
</ul>
<p><strong>TMM normalization with low counts (why an initial round of filtering is required before TMM):</strong></p>
<ul>
<li><a href="https://ltla.github.io/ChIPSeqThoughts/lowcount_norm/lowcount_norm.html" class="uri">https://ltla.github.io/ChIPSeqThoughts/lowcount_norm/lowcount_norm.html</a></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/daianna21\.github\.io\/daianna_blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="lahuuki/lahuuki.github.io" data-repo-id="MDEwOlJlcG9zaXRvcnkzNDM1NTI0NjA=" data-category="General" data-category-id="DIC_kwDOFHoxzM4CfkDR" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
<input type="hidden" id="giscus-base-theme" value="light">
<input type="hidden" id="giscus-alt-theme" value="dark">
</div> <!-- /content -->




</body></html>