<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daianna Gonzalez-Padilla">
<meta name="dcterms.date" content="2024-12-12">

<title>Daianna Gonzalez-Padilla - Exploratory Data Analysis for RNA-seq data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../images/logo-removebg.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Daianna Gonzalez-Padilla</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">About me</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blog.html"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://daianna21.github.io"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="http://www.linkedin.com/in/daianna-glez"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://x.com/daianna_glez/"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:glezdaianna@gmail.com"> <i class="bi bi-envelope-at" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../blog.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#what-youll-learn-here" id="toc-what-youll-learn-here" class="nav-link" data-scroll-target="#what-youll-learn-here">What you’ll learn here</a></li>
  <li><a href="#an-example-data-set" id="toc-an-example-data-set" class="nav-link" data-scroll-target="#an-example-data-set">An example data set</a>
  <ul class="collapse">
  <li><a href="#download-data" id="toc-download-data" class="nav-link" data-scroll-target="#download-data">Download data</a></li>
  <li><a href="#study-design-and-aims" id="toc-study-design-and-aims" class="nav-link" data-scroll-target="#study-design-and-aims">Study design and aims</a></li>
  <li><a href="#sample-variables" id="toc-sample-variables" class="nav-link" data-scroll-target="#sample-variables">Sample variables</a>
  <ul class="collapse">
  <li><a href="#main-sample-attributes" id="toc-main-sample-attributes" class="nav-link" data-scroll-target="#main-sample-attributes">Main sample attributes</a></li>
  <li><a href="#quality-control-metrics" id="toc-quality-control-metrics" class="nav-link" data-scroll-target="#quality-control-metrics">Quality-control metrics</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#exploring-count-data-todo" id="toc-exploring-count-data-todo" class="nav-link" data-scroll-target="#exploring-count-data-todo">1. Exploring count data TODO</a>
  <ul class="collapse">
  <li><a href="#read-count-distribution-todo" id="toc-read-count-distribution-todo" class="nav-link" data-scroll-target="#read-count-distribution-todo">1.1 Read count distribution TODO</a></li>
  <li><a href="#count-log-normalization-todo" id="toc-count-log-normalization-todo" class="nav-link" data-scroll-target="#count-log-normalization-todo">1.2 Count log-normalization TODO</a></li>
  <li><a href="#filtering-lowly-expressed-genes-todo" id="toc-filtering-lowly-expressed-genes-todo" class="nav-link" data-scroll-target="#filtering-lowly-expressed-genes-todo">1.3 Filtering lowly-expressed genes TODO</a></li>
  </ul></li>
  <li><a href="#filtering-low-quality-samples" id="toc-filtering-low-quality-samples" class="nav-link" data-scroll-target="#filtering-low-quality-samples">3. Filtering low-quality samples</a>
  <ul class="collapse">
  <li><a href="#explore-differences-in-qc-metrics-between-sample-groups" id="toc-explore-differences-in-qc-metrics-between-sample-groups" class="nav-link" data-scroll-target="#explore-differences-in-qc-metrics-between-sample-groups">3.1 Explore differences in QC metrics between sample groups</a></li>
  <li><a href="#explore-relationships-between-sample-variables" id="toc-explore-relationships-between-sample-variables" class="nav-link" data-scroll-target="#explore-relationships-between-sample-variables">3.2 Explore relationships between sample variables</a></li>
  <li><a href="#identify-sample-outliers-boxplots-and-pca" id="toc-identify-sample-outliers-boxplots-and-pca" class="nav-link" data-scroll-target="#identify-sample-outliers-boxplots-and-pca">3.3 Identify sample outliers (Boxplots and PCA)</a></li>
  </ul></li>
  <li><a href="#variable-associations-with-gene-expression" id="toc-variable-associations-with-gene-expression" class="nav-link" data-scroll-target="#variable-associations-with-gene-expression">4. Variable associations with gene expression</a>
  <ul class="collapse">
  <li><a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction">4.1 Dimensionality reduction</a></li>
  <li><a href="#correlation-between-variables" id="toc-correlation-between-variables" class="nav-link" data-scroll-target="#correlation-between-variables">4.2 Correlation between variables</a></li>
  <li><a href="#partition-of-gene-expression-variance" id="toc-partition-of-gene-expression-variance" class="nav-link" data-scroll-target="#partition-of-gene-expression-variance">4.2 Partition of gene expression variance</a></li>
  </ul></li>
  <li><a href="#conclusions" id="toc-conclusions" class="nav-link" data-scroll-target="#conclusions">Conclusions</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exploratory Data Analysis for RNA-seq data</h1>
  <div class="quarto-categories">
    <div class="quarto-category">TODO</div>
  </div>
  </div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Daianna Gonzalez-Padilla </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 12, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>⚠️ This page is under development.</p>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>Saying that over 50% of biological data analysis is about exploring and cleaning data is not an exaggeration. In fact, that’s the way it’s supposed to be. Biological data are inherently noisy. Not only biological differences exist between cells and samples, but all sampling and experimental procedures for data generation, introduce variability, errors, and biases into the data that we have to first spot to then clean or correct.</p>
<p>Exploratory Data Analysis (EDA) is a primordial initial step in which we get to know the overall and specific aspects of the data. This is a process where we create tons of plots to uncover and visualize the data structures, patterns, relationships, and outliers, through which we identify data issues, drive hypoteses, and guide subsequent statistical testing.</p>
<p>EDA is not a pre-established path with ordered steps to follow. What you do, what you plot depends on the particularities of the data you have and the questions you aim to answer; it is completely context dependent. Moreover, many exploratory analyses you perform depend on what previous explorations informed, adding to the global understanding of the data. However, there are basic exploratory steps that establish a foundation to move forward in your analyses.</p>
<p>In transcriptomic data analysis, EDA includes exploring the read count distribution, filtering of lowly-expressed genes and poor-quality samples/cells, determine biological and technical sources of gene expression variation, and detect confounding factors.</p>
<p>In this blog post, you will learn how to perform EDA by exploring an example gene expression data set, aiming to provide a methodological but flexible framework for you to guide your own exploratory analysis.</p>
</section>
<section id="what-youll-learn-here" class="level1">
<h1>What you’ll learn here</h1>
<ul>
<li><p>Know the type of exploratory analyses that can be done for RNA-seq data and the rationale behind them.</p></li>
<li><p>Understand the insights that can be derived from explorations and how they inform posterior analysis decisions.</p></li>
<li><p>Learn of poweful visualization tools to explore your data.</p></li>
</ul>
</section>
<section id="an-example-data-set" class="level1">
<h1>An example data set</h1>
<section id="download-data" class="level2">
<h2 class="anchored" data-anchor-id="download-data">Download data</h2>
<p>We’ll use <a href="https://rna.recount.bio"><em>recount3</em></a> <span class="citation" data-cites="leonardocollado-torres2020 wilks2021a">(<a href="#ref-leonardocollado-torres2020" role="doc-biblioref">Leonardo Collado-Torres 2020</a>; <a href="#ref-wilks2021a" role="doc-biblioref">Wilks et al. 2021</a>)</span> package to download real gene expression data from a transcriptomic study, including sample metadata and quality-control (QC) metrics.</p>
<p>The chosen study can be found in the Sequence Read Archive (<a href="https://www.ncbi.nlm.nih.gov/sra">SRA</a>) under ID <code>SRP107565</code>, and it’s titled “<strong>Multiomics Profiling Establishes the Polypharmacology of FDA-Approved CDK4/6 Inhibitors and the Potential for Differential Clinical Activity”</strong> (<a href="10.1016/j.chembiol.2019.05.005">Marc Hafner et al., 2019</a>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(recount3)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Download all available projects in human in recount3</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>human_projects <span class="ot">&lt;-</span> <span class="fu">available_projects</span>()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">## Download gene expression data and metadata from SRP107565 study</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>proj_info <span class="ot">&lt;-</span> <span class="fu">subset</span>(</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    human_projects,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    project <span class="sc">==</span> <span class="st">"SRP107565"</span> <span class="sc">&amp;</span> project_type <span class="sc">==</span> <span class="st">"data_sources"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">recount3_url =</span> <span class="st">"https://sciserver.org/public-data/recount3/data"</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>proj_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    project organism file_source     project_home project_type n_samples
1 SRP107565    human         sra data_sources/sra data_sources       216</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create RangedSummarizedExperiment object to handle RNA-seq and sample data </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>rse <span class="ot">&lt;-</span> <span class="fu">create_rse</span>(proj_info)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Gene expression data in assay(rse): for first 5 genes and 5 samples</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(rse)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  SRR5579425 SRR5579426 SRR5579433 SRR5579434 SRR5579435
ENSG00000278704.1          0          0          0          0          0
ENSG00000277400.1          0          0          0          0          0
ENSG00000274847.1          0          0          0          0          0
ENSG00000277428.1          0          0          0          0          0
ENSG00000276256.1          0          0          0          0          0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Sample metadata in colData(rse): for first 6 samples and 3 variables</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(rse)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
             rail_id external_id       study
           &lt;integer&gt; &lt;character&gt; &lt;character&gt;
SRR5579425   1000031  SRR5579425   SRP107565
SRR5579426   1000045  SRR5579426   SRP107565
SRR5579433   1000255  SRR5579433   SRP107565
SRR5579434   1000270  SRR5579434   SRP107565
SRR5579435   1000286  SRR5579435   SRP107565
SRR5579436   1000301  SRR5579436   SRP107565</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Gene data in rowData(rse): for first 6 genes and 3 variables</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(rse)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
                    source     type bp_length
                  &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
ENSG00000278704.1  ENSEMBL     gene      2237
ENSG00000277400.1  ENSEMBL     gene      2179
ENSG00000274847.1  ENSEMBL     gene      1599
ENSG00000277428.1  ENSEMBL     gene       101
ENSG00000276256.1  ENSEMBL     gene      2195
ENSG00000278198.1  ENSEMBL     gene      1468</code></pre>
</div>
</div>
</section>
<section id="study-design-and-aims" class="level2">
<h2 class="anchored" data-anchor-id="study-design-and-aims">Study design and aims</h2>
<p>In this study, the authors performed transcriptomic, proteomic, biochemical, and phenotypic assays to compare the activities of 3 inhibitors of cyclin-dependent kinases 4/6 (CDK4/6): abemaciclib, palbociclib, and ribociclib, which are used to treat hormone receptor-positive breast cancer. For our purposes, only transcriptomic data have been downloaded to be explored.</p>
<p>Specifically, in the transcriptomic experiment, 7 breast cancer human cell lines were drug-treated with <span style="color: #EE7621"><strong>0.3</strong></span>, <span style="color: #CD3700"><strong>1</strong></span> OR <span style="color: #8B2500"><strong>3</strong></span> µM of <span style="color: #0000FF"><strong>abemaciclib</strong></span>, <span style="color: #EE3B3B"><strong>palbociclib</strong></span>, OR <span style="color: #CDC673"><strong>ribociclib</strong></span> (OR untreated = <span style="color: #A6A6A6"><strong>control</strong></span> <strong>=</strong> drug concentration of <span style="color: #A6A6A6"><strong>0</strong></span> µM), and mRNA sequencing was performed after <strong>6</strong> OR <span style="color: #00CD66"><strong>24</strong></span> hrs of exposure. Including replicates, a total of 216 bulk samples were sequenced.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/study_design.png" class="img-fluid figure-img" width="363"></p>
<figcaption>Ilustration taken from <strong>Figure 1A</strong> in <a href="10.1016/j.chembiol.2019.05.005">Marc Hafner et al., 2019</a>.</figcaption>
</figure>
</div>
<p>Exploring the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html"><em>RangedSummarizedExperiment</em></a> <span class="citation" data-cites="martinmorgan2017 huber2015">(<a href="#ref-martinmorgan2017" role="doc-biblioref">Martin Morgan 2017</a>; <a href="#ref-huber2015" role="doc-biblioref">Huber et al. 2015</a>)</span> object (<code>rse</code>) we just created, we can see we have raw expression data for 63,856 genes across 216 samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(rse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 63856   216</code></pre>
</div>
</div>
<p>Processing the sample attributes contained in <code>sra.sample_attributes</code>, we can check the number of samples from each cell line, treated with each drug, at each concentration, and time. Let’s add individual columns in the sample metadata for these variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Sample attributes</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(rse<span class="sc">$</span>sra.sample_attributes, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "agent;;Abemaciclib|cell line;;BT20|dose;;0.3 uM|source_name;;BT20 breast cancer cell line|time;;6 hr"
[2] "agent;;Abemaciclib|cell line;;BT20|dose;;0.3 uM|source_name;;BT20 breast cancer cell line|time;;6 hr"
[3] "agent;;Palbociclib|cell line;;BT20|dose;;3 uM|source_name;;BT20 breast cancer cell line|time;;6 hr"  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Divide each string by "|" and extract drug</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>rse<span class="sc">$</span>Drug <span class="ot">=</span> <span class="fu">sapply</span>(rse<span class="sc">$</span>sra.sample_attributes, <span class="cf">function</span>(x){<span class="fu">strsplit</span>(<span class="fu">strsplit</span>(x, <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>], <span class="st">";;"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]}) <span class="sc">%&gt;%</span> unname</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Num of samples treated with each drug</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(rse<span class="sc">$</span>Drug)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Abemaciclib     Control Palbociclib  Ribociclib 
         84          28          52          52 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Num of samples of each cell line</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>rse<span class="sc">$</span>Cell_line <span class="ot">=</span> <span class="fu">sapply</span>(rse<span class="sc">$</span>sra.sample_attributes, <span class="cf">function</span>(x){<span class="fu">strsplit</span>(<span class="fu">strsplit</span>(x, <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>], <span class="st">";;"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]}) <span class="sc">%&gt;%</span> unname</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(rse<span class="sc">$</span>Cell_line)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   BT20   BT549 HCC1419 HCC1806  HS578T    MCF7    T47D 
     32      32      32      32      32      24      32 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Num of samples treated with each dose</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>rse<span class="sc">$</span>Concentration <span class="ot">=</span>  <span class="fu">sapply</span>(rse<span class="sc">$</span>sra.sample_attributes, <span class="cf">function</span>(x){<span class="fu">strsplit</span>(<span class="fu">strsplit</span>(x, <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>)[[<span class="dv">1</span>]][<span class="dv">3</span>], <span class="st">";;"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]}) <span class="sc">%&gt;%</span> unname</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(rse<span class="sc">$</span>Concentration)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  0 uM 0.3 uM   1 uM   3 uM 
    28     80      4    104 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Num of samples sequenced after each time</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>rse<span class="sc">$</span>Time <span class="ot">=</span>  <span class="fu">sapply</span>(rse<span class="sc">$</span>sra.sample_attributes, <span class="cf">function</span>(x){<span class="fu">strsplit</span>(<span class="fu">strsplit</span>(x, <span class="st">"</span><span class="sc">\\</span><span class="st">|"</span>)[[<span class="dv">1</span>]][<span class="dv">5</span>], <span class="st">";;"</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]}) <span class="sc">%&gt;%</span> unname</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(rse<span class="sc">$</span>Time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
24 hr  6 hr 
  108   108 </code></pre>
</div>
</div>
</section>
<section id="sample-variables" class="level2">
<h2 class="anchored" data-anchor-id="sample-variables">Sample variables</h2>
<p>In any EDA, we first need to know what the sample variables in the metadata stand for in order to interpret accurately any insights derived from them. A second important consideration is that, even when there are well-established QC metrics used to filter samples, exploring a broader set of them is convenient to find unexpected sample quality differences.</p>
<p>We start by defining the sample-level variables we’ll explore in our EDA.</p>
<section id="main-sample-attributes" class="level3">
<h3 class="anchored" data-anchor-id="main-sample-attributes">Main sample attributes</h3>
<ul>
<li><p><code>Cell_line</code>: name of breast cancer cell line.</p></li>
<li><p><code>Drug</code>: the drug with which a given cell line was treated.</p></li>
<li><p><code>Concentration</code>: the concentration of the drug administered to the cell line.</p></li>
<li><p><code>Time</code>: drug exposure time for each cell line.</p></li>
</ul>
</section>
<section id="quality-control-metrics" class="level3">
<h3 class="anchored" data-anchor-id="quality-control-metrics">Quality-control metrics</h3>
<p>QC metrics are measurements of a sample’s RNA composition and the mapping of its reads to a reference genome, reflecting the integrity of the cells, and the quality of the mRNA extraction, library preparation, sequencing, and read alignment steps.</p>
<p>The following metrics were collected by <em>recount3</em> through a number of tools.</p>
<ul>
<li><p>Alignment-related metrics obtained through <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR</a> <span class="citation" data-cites="dobin2012">(<a href="#ref-dobin2012" role="doc-biblioref">Dobin et al. 2012</a>)</span>:</p>
<ul>
<li><p><strong><code>all_mapped_reads</code></strong>: total number of aligned reads aligned.</p></li>
<li><p><strong><code>uniquely_mapped_reads_%</code></strong>: number of reads that mapped to a single locus, of the total input reads.</p></li>
<li><p><strong><code>uniquely_mapped_reads_number</code></strong>: number of reads that mapped to a single locus.</p></li>
<li><p><strong><code>%_of_reads_mapped_to_multiple_loci</code></strong>: number of reads that mapped to multiple loci, of the input reads.</p></li>
<li><p><strong><code>number_of_reads_mapped_to_multiple_loci</code></strong>: number of reads that mapped to multiple loci.</p></li>
<li><p><strong><code>%_of_reads_unmapped:_other</code></strong>: reads that didn’t map due to no acceptable seed/windows, of all input reads.</p></li>
<li><p><strong><code>number_of_reads_unmapped:_other</code></strong>: number of reads left unmapped due to no acceptable seed/windows.</p></li>
<li><p><strong><code>%_of_reads_unmapped:_too_many_mismatches</code></strong>: Number of reads where best alignment has more mismatches than max allowed number of mismatches divided by number of input reads</p></li>
<li><p><strong><code>number_of_reads_unmapped:_too_many_mismatches</code></strong>: Number of reads where best alignment has more mismatches than max allowed number of mismatches</p></li>
<li><p><strong><code>%_of_reads_unmapped:_too_short</code></strong>: Number of reads where best alignment was shorter than min allowed mapped length divided by number of input reads.</p></li>
<li><p><strong><code>number_of_reads_unmapped:_too_short</code></strong>: Number of reads where best alignment was shorter than min allowed mapped length.</p></li>
</ul></li>
<li><p><a href="http://www.htslib.org">SAMtools</a> <span class="citation" data-cites="danecek2021">(<a href="#ref-danecek2021" role="doc-biblioref">Danecek et al. 2021</a>)</span> statistics for chromosome-specific read mapping. Of special interest are the reads mapping to the mitochondrial and sex chromosomes (see explanation in Step X):</p>
<ul>
<li><p><strong><code>aligned_reads%.chrm</code></strong>: Percent of reads aligning to the mitochondrial genome.</p></li>
<li><p><strong><code>aligned_reads%.chrx</code></strong>: Percent of reads aligning to chromosome X.</p></li>
<li><p><strong><code>aligned_reads%.chry</code></strong>: Precent of reads aligning to chromosome Y.</p></li>
</ul></li>
<li><p>Sequencing coverage summaries in gene annotations by <a href="https://github.com/ChristopherWilks/megadepth">megadepth</a> (<code>bc</code>) <span class="citation" data-cites="wilks2020">(<a href="#ref-wilks2020" role="doc-biblioref">Wilks et al. 2020</a>)</span> and <a href="https://subread.sourceforge.net/featureCounts.html"><em>featureCounts</em></a> <span class="citation" data-cites="liao2013">(<a href="#ref-liao2013" role="doc-biblioref">Liao, Smyth, and Shi 2013</a>)</span> (<code>fc</code>):</p>
<ul>
<li><p><strong><code>bc_auc.all_reads_all_bases</code></strong>: Area under coverage (total depth of coverage evaluated at all bases) for all alignments</p></li>
<li><p><strong><code>bc_auc.all_reads_annotated_bases</code></strong>: Area under coverage for all alignments, but only for bases in annotated exons</p></li>
<li><p><strong><code>bc_auc.unique_reads_all_bases</code></strong>: Area under coverage for uniquely aligned reads</p></li>
<li><p><strong><code>bc_auc.unique_reads_annotated_bases</code></strong>: Area under coverage for uniquely aligned reads, but only for bases in annotated exons.</p></li>
<li><p><strong><code>exon_fc_count_all.total</code></strong>: Total number of fragments, including multi-mappers, input to&nbsp;<code>featureCounts</code></p></li>
<li><p><strong><code>exon_fc_count_all.assigned</code></strong>: Number of fragments, including multi-mappers, assigned by&nbsp;<code>featureCounts</code>&nbsp;to an exon</p></li>
<li><p><strong><code>exon_fc_count_unique.total</code></strong>: Total number of uniquely mapping fragments input to&nbsp;<code>featureCounts</code></p></li>
<li><p><strong><code>exon_fc_count_unique.assigned</code></strong>: Number of uniquely mapping fragments assigned by&nbsp;<code>featureCounts</code>&nbsp;to an exon.</p></li>
<li><p><strong><code>gene_fc_count_all.total</code></strong>: Total number of fragments, including multi-mappers, input to&nbsp;<code>featureCounts</code></p></li>
<li><p><strong><code>gene_fc_count_all.assigned</code></strong>: Number of fragments, including multi-mappers, assigned by&nbsp;<code>featureCounts</code>&nbsp;to a gene</p></li>
<li><p><strong><code>gene_fc_count_unique.total</code></strong>: Total number of uniquely mapping fragments input to&nbsp;<code>featureCounts</code></p></li>
<li><p><strong><code>gene_fc_count_unique.assigned</code></strong>: Number of uniquely mapping fragments assigned by&nbsp;<code>featureCounts</code>&nbsp;to a gene.</p></li>
</ul></li>
</ul>
<p>For more details on the computation of these metrics please refer to <em>recount3</em> <a href="http://rna.recount.bio/docs/quality-check-fields.html">documentation</a> and the manual of each tool.</p>
</section>
</section>
</section>
<section id="exploring-count-data-todo" class="level1">
<h1>1. Exploring count data TODO</h1>
<section id="read-count-distribution-todo" class="level2">
<h2 class="anchored" data-anchor-id="read-count-distribution-todo">1.1 Read count distribution TODO</h2>
<p>A first thing to plot is the frequency of the raw counts across all genes and samples, before any transformation. This gives us an idea of how sparse the data are, i.e., if a large percentage of the counts are zero.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Raw counts</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">raw_counts =</span> <span class="fu">as.vector</span>(<span class="fu">assays</span>(rse)<span class="sc">$</span>raw_counts))</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data, <span class="fu">aes</span>(<span class="at">x =</span> raw_counts)) <span class="sc">+</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_histogram</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"lightgray"</span>) <span class="sc">+</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Raw counts"</span>, <span class="at">y =</span> <span class="st">"Frecuency"</span>) <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_classic</span>()</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>plot <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.margin =</span> <span class="fu">unit</span>(<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">4</span>), <span class="st">"cm"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/hist_raw_counts-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Percentage of counts that are zero</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">which</span>(<span class="fu">as.vector</span>(<span class="fu">assays</span>(rse)<span class="sc">$</span>raw_counts) <span class="sc">==</span> <span class="dv">0</span>)) <span class="sc">/</span> <span class="fu">length</span>(<span class="fu">as.vector</span>(<span class="fu">assays</span>(rse)<span class="sc">$</span>raw_counts)) <span class="sc">*</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 61.05911</code></pre>
</div>
</div>
<p>More than 60% of the total counts across all genes and bulk samples are 0. Even though this is a large fraction, it is not surprising as we don’t expect most genes to be expressed in every condition; scRNA-seq data is even more sparse. In <strong>step 1.3</strong> we’ll deal with lowly-expressed genes …</p>
</section>
<section id="count-log-normalization-todo" class="level2">
<h2 class="anchored" data-anchor-id="count-log-normalization-todo">1.2 Count log-normalization TODO</h2>
</section>
<section id="filtering-lowly-expressed-genes-todo" class="level2">
<h2 class="anchored" data-anchor-id="filtering-lowly-expressed-genes-todo">1.3 Filtering lowly-expressed genes TODO</h2>
<p>There are multiple approaches to remove zero- and lowly-expressed genes. Here we’ll adopt the approach of removing genes with less than X CPM in at least X % samples …</p>
<p><a href="https://carpentries-incubator.github.io/rna-seq-data-for-ml/episode5.html" class="uri">https://carpentries-incubator.github.io/rna-seq-data-for-ml/episode5.html</a></p>
</section>
</section>
<section id="filtering-low-quality-samples" class="level1">
<h1>3. Filtering low-quality samples</h1>
<section id="explore-differences-in-qc-metrics-between-sample-groups" class="level2">
<h2 class="anchored" data-anchor-id="explore-differences-in-qc-metrics-between-sample-groups">3.1 Explore differences in QC metrics between sample groups</h2>
<p>Explore sex differences: sexual chr % can help to confirm the sex of the donors based on alignments to sex chromosomes.</p>
<p><a href="https://frontlinegenomics.com/how-to-ngs-quality-control/" class="uri">https://frontlinegenomics.com/how-to-ngs-quality-control/</a></p>
<p><a href="https://www.nature.com/articles/s41592-023-01946-4" class="uri">https://www.nature.com/articles/s41592-023-01946-4</a></p>
</section>
<section id="explore-relationships-between-sample-variables" class="level2">
<h2 class="anchored" data-anchor-id="explore-relationships-between-sample-variables">3.2 Explore relationships between sample variables</h2>
</section>
<section id="identify-sample-outliers-boxplots-and-pca" class="level2">
<h2 class="anchored" data-anchor-id="identify-sample-outliers-boxplots-and-pca">3.3 Identify sample outliers (Boxplots and PCA)</h2>
</section>
</section>
<section id="variable-associations-with-gene-expression" class="level1">
<h1>4. Variable associations with gene expression</h1>
<p>Identify technical sources of data variability to then reduce the systematic noise from the data.</p>
<section id="dimensionality-reduction" class="level2">
<h2 class="anchored" data-anchor-id="dimensionality-reduction">4.1 Dimensionality reduction</h2>
</section>
<section id="correlation-between-variables" class="level2">
<h2 class="anchored" data-anchor-id="correlation-between-variables">4.2 Correlation between variables</h2>
</section>
<section id="partition-of-gene-expression-variance" class="level2">
<h2 class="anchored" data-anchor-id="partition-of-gene-expression-variance">4.2 Partition of gene expression variance</h2>
</section>
</section>
<section id="conclusions" class="level1">
<h1>Conclusions</h1>
<p>Answering these questions we may be able to formulate&nbsp;<strong>hypothesis</strong>&nbsp;that could guide future analysis or experiments.</p>
<p>EDA is extremely illuminating when performed comprehensively and you may find an answer to your data issues or lack of signals by looking back to this step.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-danecek2021" class="csl-entry" role="listitem">
Danecek, Petr, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, et al. 2021. <span>“Twelve Years of SAMtools and BCFtools.”</span> <em>GigaScience</em> 10 (2). <a href="https://doi.org/10.1093/gigascience/giab008">https://doi.org/10.1093/gigascience/giab008</a>.
</div>
<div id="ref-dobin2012" class="csl-entry" role="listitem">
Dobin, Alexander, Carrie A. Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R. Gingeras. 2012. <span>“STAR: Ultrafast Universal RNA-Seq Aligner.”</span> <em>Bioinformatics</em> 29 (1): 15–21. <a href="https://doi.org/10.1093/bioinformatics/bts635">https://doi.org/10.1093/bioinformatics/bts635</a>.
</div>
<div id="ref-huber2015" class="csl-entry" role="listitem">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. <span>“Orchestrating High-Throughput Genomic Analysis with Bioconductor.”</span> <em>Nature Methods</em> 12 (2): 115–21. <a href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>.
</div>
<div id="ref-leonardocollado-torres2020" class="csl-entry" role="listitem">
Leonardo Collado-Torres. 2020. <span>“Recount3.”</span> <a href="https://doi.org/10.18129/B9.BIOC.RECOUNT3">https://doi.org/10.18129/B9.BIOC.RECOUNT3</a>.
</div>
<div id="ref-liao2013" class="csl-entry" role="listitem">
Liao, Yang, Gordon K. Smyth, and Wei Shi. 2013. <span>“featureCounts: An Efficient General Purpose Program for Assigning Sequence Reads to Genomic Features.”</span> <em>Bioinformatics</em> 30 (7): 923–30. <a href="https://doi.org/10.1093/bioinformatics/btt656">https://doi.org/10.1093/bioinformatics/btt656</a>.
</div>
<div id="ref-martinmorgan2017" class="csl-entry" role="listitem">
Martin Morgan, Valerie Obenchain. 2017. <span>“SummarizedExperiment.”</span> <a href="https://doi.org/10.18129/B9.BIOC.SUMMARIZEDEXPERIMENT">https://doi.org/10.18129/B9.BIOC.SUMMARIZEDEXPERIMENT</a>.
</div>
<div id="ref-wilks2020" class="csl-entry" role="listitem">
Wilks, Christopher, Omar Ahmed, Daniel N. Baker, David Zhang, Leonardo Collado-Torres, and Ben Langmead. 2020. <span>“Megadepth: Efficient Coverage Quantification for BigWigs and BAMs.”</span> <a href="http://dx.doi.org/10.1101/2020.12.17.423317">http://dx.doi.org/10.1101/2020.12.17.423317</a>.
</div>
<div id="ref-wilks2021a" class="csl-entry" role="listitem">
Wilks, Christopher, Shijie C. Zheng, Feng Yong Chen, Rone Charles, Brad Solomon, Jonathan P. Ling, Eddie Luidy Imada, et al. 2021. <span>“Recount3: Summaries and Queries for Large-Scale RNA-Seq Expression and Splicing.”</span> <em>Genome Biology</em> 22 (1). <a href="https://doi.org/10.1186/s13059-021-02533-6">https://doi.org/10.1186/s13059-021-02533-6</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/daianna21\.github\.io\/daianna_blog\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="lahuuki/lahuuki.github.io" data-repo-id="MDEwOlJlcG9zaXRvcnkzNDM1NTI0NjA=" data-category="General" data-category-id="DIC_kwDOFHoxzM4CfkDR" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
<input type="hidden" id="giscus-base-theme" value="light">
<input type="hidden" id="giscus-alt-theme" value="dark">
</div> <!-- /content -->




</body></html>